{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "from os.path import join\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "from biolabsim import Host, Strain, Ecol\n",
    "from biolabsim import measure_EnzymeLevel1, Help_GenomeGenerator, Strain, make_UpdateExpression, Help_StrainCharacterizer, Help_PromoterStrength, Help_Expr2Flux\n",
    "from Bio.Seq import Seq\n",
    "from pandas.core.frame import DataFrame\n",
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TranscriptionFactor(df, RegID: str, act_rep: str, PromSeq: str, ORFSeq: str):\n",
    "    newdf = df\n",
    "\n",
    "    # check if TF is already present, add columns to the default TF accordingly, no zero initiation of RegType column\n",
    "    if any('TF_regulated' in s for s in newdf.columns.values):\n",
    "        Flux = 10\n",
    "        Expr = Help_PromoterStrength('Ecol', PromSeq)\n",
    "        TF = {\n",
    "            'RctID': [RegID],\n",
    "            'Expression': [Expr],  # needs to be != 0\n",
    "            'Promoter': [PromSeq],  # better be unique\n",
    "            'ORF': [ORFSeq],  # must be unique\n",
    "            'Fluxes': [Flux],  # value not further relevant\n",
    "            'Expr2Flux':  [Help_Expr2Flux(Flux,Expr)],\n",
    "            'TF_regulated': [0],\n",
    "            'RegType': act_rep\n",
    "        }\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "\n",
    "    else:\n",
    "        # add TF row:\n",
    "        Flux = 10\n",
    "        Expr = Help_PromoterStrength('Ecol', PromSeq)\n",
    "        TF = {\n",
    "            'RctID': [RegID],\n",
    "            'Expression': [Expr],\n",
    "            'Promoter': [PromSeq],\n",
    "            'ORF': [ORFSeq],\n",
    "            'Fluxes': [Flux],\n",
    "            'Expr2Flux': [Help_Expr2Flux(Flux,Expr)]\n",
    "        }\n",
    "\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "\n",
    "        # add TF_regulated and RegType column:\n",
    "        zeroarray = np.zeros(len(newdf), dtype=int)\n",
    "        newdf = newdf.assign(TF_regulated=zeroarray)  \n",
    "        newdf = newdf.assign(RegType=zeroarray)\n",
    "\n",
    "        TF_Indx = (newdf['RctID'] == RegID)  # access TF row\n",
    "        newdf.loc[TF_Indx, 'RegType'] = act_rep  \n",
    "        \n",
    "    return newdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TranscriptionFactorReg(df, RegID: str, EnzymeTarget: list):\n",
    "    newdf = df\n",
    "    # set Regulation of some Proteins:\n",
    "    for ETarget in EnzymeTarget:\n",
    "        # Future: fct recognizes what substrates are there, then makes TF_act or TF_rep\n",
    "        # define new and changed functions\n",
    "\n",
    "        myIndx = (newdf['RctID'] == ETarget)\n",
    "        newdf.loc[myIndx, 'TF_regulated'] = RegID\n",
    "\n",
    "    return newdf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Update_DF(df):\n",
    "    # update expression and Expr2Flux\n",
    "    corrExpr = calculate_StandardRegulatedExpression(df)\n",
    "    df['Expression'] = corrExpr\n",
    "    # update for Expression to Flux correlation, for all a linear factor\n",
    "    myExpr2Flux = Help_Expr2Flux(df['Fluxes'], corrExpr)\n",
    "    df['Expr2Flux'] = myExpr2Flux\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Help_mutatePromoter(Mutant_Targets: list, wtStrain, mtStrain, insertSeq: str, cutSeq: str ='ATTGA'):\n",
    "    # Mutants to be generated\n",
    "\n",
    "    for myMutant in Mutant_Targets:\n",
    "\n",
    "        MTdf = mtStrain.genes_df.copy()\n",
    "        Mutant_Indx = tuple(MTdf[MTdf['RctID'] == myMutant].index.values)\n",
    "        # Finding the index for later printing\n",
    "        WTIndx = str(wtStrain.genome).find(wtStrain.genes_df.loc[Mutant_Indx, 'Promoter'])\n",
    "    #     # Generating target promoter sequence, ATTGA is always there\n",
    "        PromTar = MTdf.loc[Mutant_Indx, 'Promoter'].replace(cutSeq,insertSeq)\n",
    "        # converting Biopython Seq class to string. This makes string replacements easier\n",
    "        MutGenome = str(mtStrain.genome)\n",
    "        MutGenome = MutGenome.replace(MTdf.loc[Mutant_Indx, 'Promoter'], PromTar)\n",
    "        MTdf.loc[Mutant_Indx, 'Promoter'] = PromTar\n",
    "    #     print('{}, Reference:\\t{},\\nMutated:\\t{}\\nwith index {}'.format(myMutant, wtHost.strain.genome[Mutant_Indx:Mutant_Indx+41], myHost.strain.genome[Mutant_Indx:Mutant_Indx+41], Mutant_Indx))\n",
    "        print(myMutant, WTIndx)\n",
    "    \n",
    "\n",
    "        \n",
    "    return MutGenome, MTdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_DetectRegulatorPromoterMut(StrainMut, RctNewDF):\n",
    "    #     for obj in (StrainWt, StrainMut):\n",
    "    #         assert isinstance(obj, Strain), 'wrong strain type: must be Strain'\n",
    "    #     assert isinstance(RctNewDF, DataFrame), 'ReactNewDF of wrong type: input dataframe'\n",
    "    GenesDF = StrainMut.genes_df\n",
    "    # finding regulators in GenesDF\n",
    "    myRegIndx = GenesDF[GenesDF['RegType'] != 0].index.values\n",
    "    # finding index of regulators with changes in the promoter\n",
    "    RegUpdatePromIndx = myRegIndx[RctNewDF.loc[myRegIndx, 'RctFlag'].values]\n",
    "    # finding name of regulators with changes in the promoter\n",
    "    RegUpdatePromName = RctNewDF.loc[RegUpdatePromIndx, 'RctID'].values\n",
    "    # finding enzymes regulated by regulator\n",
    "    EnzUpdateReguIndx = GenesDF['TF_regulated'].isin(RegUpdatePromName)\n",
    "\n",
    "    return EnzUpdateReguIndx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_EnzymeLevel(HostName: str, StrainWT, StrainMut):\n",
    "    '''\n",
    "    The function differences of expression levels of enzymes between two strains.\n",
    "    '''\n",
    "    RefGenDF = StrainWT.genes_df\n",
    "    RefGenome = str(StrainWT.genome)\n",
    "    MutGenome = str(StrainMut.genome)\n",
    "    RefModel = StrainWT.model\n",
    "\n",
    "    RctNewDF = Help_StrainCharacterizer(HostName, RefGenDF, RefGenome, MutGenome, RefModel)\n",
    "\n",
    "    # identifying changes in regulator expression\n",
    "    RegNewAr = make_DetectRegulatorPromoterMut(StrainMut, RctNewDF)\n",
    "\n",
    "    # combining promoter changes enzyme + regulator\n",
    "    AllNewAr = RctNewDF['RctFlag'].values | RegNewAr\n",
    "    AllNewDF = deepcopy(RctNewDF)\n",
    "    AllNewDF['RctFlag'] = AllNewAr\n",
    "\n",
    "    return AllNewDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_StandardRegulatedExpression(WTdf):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf['TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "    print(RegTypes)\n",
    "\n",
    "\n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "             \n",
    "        MaxExpr = 2 * WTdf.loc[EnzIndx, 'Expression'].values  # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "        cTF = WTdf.loc[RegIndx, 'Expression'].values  # abhängig von geändertem TF-Promotor\n",
    "        K = WTdf.loc[EnzIndx, 'Expression'].values  # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95  # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.\n",
    "        #         print ('Eqn: ', RegType, MaxExpr, cTF, K)\n",
    "\n",
    "        if EquType == 'Activator':\n",
    "            corrExpr = MaxExpr / (1 + (K / cTF) ** n)\n",
    "\n",
    "        elif EquType == 'Repressor':\n",
    "            corrExpr = MaxExpr * (1 - 1 / (1 + (K / cTF) ** n))\n",
    "\n",
    "        PreProcFlux = deepcopy(WTdf['Expression'].values)\n",
    "        PreProcFlux[EnzIndx] = corrExpr\n",
    "\n",
    "    return PreProcFlux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_mutatedRegulatedExpression(WTdf, AllNewDF):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf.loc[AllNewDF['RctFlag'], 'TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "\n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "        \n",
    "        MaxExpr = 2 * AllNewDF.loc[EnzIndx, 'RefExpr'].values  # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "        cTF = AllNewDF.loc[RegIndx, 'NewExpr'].values  # abhängig von geändertem TF-Promotor\n",
    "        K = AllNewDF.loc[EnzIndx, 'RefExpr'].values  # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95  # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.\n",
    "        #         print ('Eqn: ', AllNewDF)\n",
    "\n",
    "        if EquType == 'Activator':\n",
    "            corrExpr = MaxExpr / (1 + (K / cTF) ** n)\n",
    "\n",
    "        elif EquType == 'Repressor':\n",
    "            corrExpr = MaxExpr * (1 - 1 / (1 + (K / cTF) ** n))\n",
    "\n",
    "        PreProcFlux = AllNewDF['NewExpr'].values\n",
    "        PreProcFlux[EnzIndx] = corrExpr\n",
    "\n",
    "    return PreProcFlux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setboundary(df, AllNewDF):\n",
    "    '''\n",
    "    Finding and setting the boundaries for cobrapy.\n",
    "    '''\n",
    "\n",
    "    # finding reactions for which the expression has changed\n",
    "    OnlyEnzIndx = df['RegType'] == 0\n",
    "    RctNewDF = AllNewDF.loc[OnlyEnzIndx, ['RctFlag']] == True\n",
    "    RctNew = np.arange((len(RctNewDF)))[np.ravel(RctNewDF.values)]\n",
    "    #     OnlyEnzIndx = WTdf[WTdf['RegType'] == 0].index.values\n",
    "    #     RctNew = AllNewDF[AllNewDF.loc[OnlyEnzIndx, 'RctFlag'] == True].index.values\n",
    "\n",
    "    # For reactions with reduced and positive flux: reduce the upper limit,\n",
    "    # For reactions with increased and positive flux: increase the lower limit\n",
    "    # For reactions with negative flux the limits are exchanged.\n",
    "    FluxPos = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux'] > 0])]\n",
    "    FluxNeg = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux'] < 0])]\n",
    "    # Finding increased and decreased fluxes\n",
    "    FluxInc = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values > 1]\n",
    "    FluxDec = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values < 1]\n",
    "\n",
    "    # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "    PosIncInd = np.intersect1d(FluxPos, FluxInc)\n",
    "    # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "    PosDecInd = np.intersect1d(FluxPos, FluxDec)\n",
    "    # Comb.3: negative flux with increased expression -> decreasing lower bound\n",
    "    NegIncInd = np.intersect1d(FluxNeg, FluxInc)\n",
    "    # Comb.4: positive flux with increased expression -> increasing lower bound\n",
    "    NegDecInd = np.intersect1d(FluxNeg, FluxDec)\n",
    "\n",
    "    Expr_Change = {'increase': np.hstack([PosIncInd, NegIncInd]), 'decrease': np.hstack([PosDecInd, NegDecInd])}\n",
    "    Set_Boundary = {'lower': np.hstack([PosIncInd, NegDecInd]), 'upper': np.hstack([PosDecInd, NegIncInd])}\n",
    "\n",
    "    return Set_Boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding reactions to change. Three possibilities exits: 1. the promoter of the enzyme itself has changed, 2. the promoter of a regulator has changed, 3. the enzyme promoter and the regulator promoter have changed\n",
    "# redefining Help_FluxCalculator in simulation/metabolism\n",
    "def Help_FluxCalculator(HostName: str, Strain, AllNewDF=None):\n",
    "    '''\n",
    "    Calculation of flux values.\n",
    "\n",
    "    This method can work in 2 modes:\n",
    "      [Strain only] : The metabolic model of the WT strain is used for calculation.\n",
    "      [Strain + AllNewDF] : An additional step of resetting boundaries is done on the model\n",
    "        before the fluxes are calculated.\n",
    "\n",
    "    TODO: The \"reset boundary\" step is mutating the `StrainMut.model`, mutation might not be intended.\n",
    "      Because of this, perhaps the other `Strain.model` gets mutated in the process.\n",
    "    '''\n",
    "    #     from ..measurement.fluxes import measure_EnzymeLevel1\n",
    "\n",
    "    # adding flux values\n",
    "    # setup of flux boundaries. For the reference boundary changes are set to 'False',\n",
    "    # for mutant strains, ractions with altered promoter sequence will change enzyme levels and boundaries must be changed accordingly, their variable is 'True'\n",
    "\n",
    "    if AllNewDF is not None:\n",
    "        print('resetting boundaries')\n",
    "        # finding reactions for which the expression has changed, and finding the new level\n",
    "        RefFlux = AllNewDF['RefFlux']\n",
    "        NewFlux = AllNewDF['NewFlux']\n",
    "\n",
    "        Set_Boundary = setboundary(Strain.genes_df, AllNewDF)\n",
    "        # Defining the model with the two combinations of either\n",
    "        # increasing lower bound (increased forward, decreased reverse reaction)\n",
    "        # decreasing upper bound (decreased forward, increased reverse reaction)\n",
    "        with Strain.model as myModel:\n",
    "            # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "            for Indx in Set_Boundary['lower']:\n",
    "                myModel.reactions[Indx].lower_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "            # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "            for Indx in Set_Boundary['upper']:\n",
    "                myModel.reactions[Indx].upper_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "\n",
    "            Fluxes = myModel.optimize()\n",
    "\n",
    "    else:\n",
    "        Fluxes = Strain.model.optimize()\n",
    "\n",
    "    return Fluxes.fluxes.values, Fluxes.objective_value"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
