{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "6wxIUGOEReDX"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System ready\n"
     ]
    }
   ],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "from os.path import join\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# loading cobrapy, a library dedicated to the analysis of genome scale metabolic models\n",
    "# from cobra.io import read_sbml_model, write_sbml_model, load_matlab_model\n",
    "\n",
    "# loading escher for metabolic network visualization\n",
    "# import escher\n",
    "# from escher import Builder\n",
    "# from time import sleep\n",
    "# escher.rc['never_ask_before_quit'] = True\n",
    "# list of available maps\n",
    "# print(escher.list_available_maps())\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "# from memote import test_model, snapshot_report\n",
    "\n",
    "from biolabsim import Host, Strain, Ecol\n",
    "\n",
    "print('System ready')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Host: Ecol\n",
      "Resources: 40\n",
      "Substrate: None\n",
      "Strain: WT\n",
      "Wild type growth rate: 0.87\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "resetting boundaries\n",
      "Host: Ecol\n",
      "Resources: 40\n",
      "Substrate: None\n",
      "Strain: Mut1\n",
      "Mutation growth rate: 0.87\n"
     ]
    }
   ],
   "source": [
    "wtHost = Ecol()\n",
    "wtHost.show_BiotechSetting()\n",
    "print('Wild type growth rate: {:.2f}'.format(wtHost.strain.objective))\n",
    "print()\n",
    "myHost = wtHost.clone_with_mutation('Mut1')\n",
    "myHost.show_BiotechSetting()\n",
    "print('Mutation growth rate: {:.2f}'.format(myHost.strain.objective))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ändern des Dataframes\n",
    "#### Hinzufügen des Transkriptionsfaktors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TranscriptionFactor(df):\n",
    "    newdf = df\n",
    "    \n",
    "    # add TF row:\n",
    "    TF = {\n",
    "    'RctID':['TF'], \n",
    "    'Expression':[2], # noch zu ändern mit Help_PromoterStrength()?\n",
    "    'Promoter':['GCCCATTGACCATACCGGAAGGAGATAATAAAGTTGCACG'],  # noch zu ändern (echt, random oder Fkt?)\n",
    "    'ORF': ['ATGGAGATGAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "    'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "    'Expr2Flux': [4.5] # noch zu ändern mit Help_Expr2Flux()?\n",
    "    }\n",
    "    dfTF = pd.DataFrame(TF)\n",
    "    newdf = newdf.append(dfTF, ignore_index=True)\n",
    "    \n",
    "    # add TF_regulated column:\n",
    "    zeroarray = np.zeros(len(newdf),dtype=int) # später zu 1 oder 0 ändern (random oder festgesetzt?)\n",
    "    newdf = newdf.assign(TF_regulated=zeroarray)\n",
    "    # statt 1 'TF' (RctID)\n",
    "    \n",
    "    # set TF as always TF_regulated:\n",
    "    filt = (newdf['RctID']=='TF') # access only TF (da index sich ändern könnte, der etwas kompliziertere Weg)\n",
    "    newdf.loc[filt, 'TF_regulated'] = 1 # TF immer TF_regulated\n",
    "\n",
    "    return newdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>TF_regulated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>1.441</td>\n",
       "      <td>GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG</td>\n",
       "      <td>ATGGGAACAAATTGA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>5.189023</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.385</td>\n",
       "      <td>GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG</td>\n",
       "      <td>ATGCCAGTGTGCTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>1.123</td>\n",
       "      <td>GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG</td>\n",
       "      <td>ATGTTTATTGAATAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>4.328461</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.813</td>\n",
       "      <td>GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG</td>\n",
       "      <td>ATGTATCAAATTTAA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-19.709134</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.457</td>\n",
       "      <td>GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG</td>\n",
       "      <td>ATGATTAGCTTCTAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>3.404245</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NADTRHD</td>\n",
       "      <td>2.126</td>\n",
       "      <td>GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG</td>\n",
       "      <td>ATGTACAATGAATAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>2.358</td>\n",
       "      <td>GCCCATTGACATTGCTGTCTAGATAGTGTAGTTTTGCACG</td>\n",
       "      <td>ATGTGTTCGAACTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>2.020916</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>3.341</td>\n",
       "      <td>GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG</td>\n",
       "      <td>ATGCCTCCCAGCTGA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>6.524841</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>1.756</td>\n",
       "      <td>GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG</td>\n",
       "      <td>ATGGGTCGCGCTTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>5.286180</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TF</td>\n",
       "      <td>2.000</td>\n",
       "      <td>GCCCATTGACCATACCGGAAGGAGATAATAAAGTTGCACG</td>\n",
       "      <td>ATGGAGATGAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RctID  Expression                                  Promoter  \\\n",
       "0       PFK       1.441  GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG   \n",
       "1       PFL       1.385  GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG   \n",
       "2       PGI       1.123  GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG   \n",
       "3       PGK       0.813  GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG   \n",
       "4       PGL       1.457  GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG   \n",
       "..      ...         ...                                       ...   \n",
       "91  NADTRHD       2.126  GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG   \n",
       "92     NH4t       2.358  GCCCATTGACATTGCTGTCTAGATAGTGTAGTTTTGCACG   \n",
       "93      O2t       3.341  GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG   \n",
       "94      PDH       1.756  GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG   \n",
       "95       TF       2.000  GCCCATTGACCATACCGGAAGGAGATAATAAAGTTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux  TF_regulated  \n",
       "0   ATGGGAACAAATTGA   7.477382   5.189023             0  \n",
       "1   ATGCCAGTGTGCTAA   0.000000   0.000000             0  \n",
       "2   ATGTTTATTGAATAA   4.860861   4.328461             0  \n",
       "3   ATGTATCAAATTTAA -16.023526 -19.709134             0  \n",
       "4   ATGATTAGCTTCTAA   4.959985   3.404245             0  \n",
       "..              ...        ...        ...           ...  \n",
       "91  ATGTACAATGAATAA   0.000000   0.000000             0  \n",
       "92  ATGTGTTCGAACTAA   4.765319   2.020916             0  \n",
       "93  ATGCCTCCCAGCTGA  21.799493   6.524841             0  \n",
       "94  ATGGGTCGCGCTTAA   9.282533   5.286180             0  \n",
       "95  ATGGAGATGAAGTAA   9.000000   4.500000             1  \n",
       "\n",
       "[96 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newdf = wtHost.strain.genes_df\n",
    "newdf = add_TranscriptionFactor(newdf)\n",
    "newdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unnötig but nett to know: random alle Gene durch TF regulieren lassen oder nicht\n",
    "\n",
    "# import random\n",
    "\n",
    "# rand_onoff = random.randint(0,1)\n",
    "    \n",
    "# newdf['TF_regulated'] = newdf['TF_regulated'].apply(lambda x: x * 0 + onoff)\n",
    "\n",
    "# newdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regulationsgleichung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-7-155daad20ded>, line 5)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-7-155daad20ded>\"\u001b[1;36m, line \u001b[1;32m5\u001b[0m\n\u001b[1;33m    EquationParameter: 'TF_regulated', cTF, 'Expression'\u001b[0m\n\u001b[1;37m                                     ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# myEquations=['linear','Hill']\n",
    "\n",
    "myModel.reactions[Indx].lower_bound = FluxEquation # (equationtype = myEquations[RctNewDF.loc[Indx, 'Expr2Flux']])\n",
    "\n",
    "EquationParameter: 'TF_regulated', cTF, 'Expression'\n",
    "\n",
    "def FluxEquation(index, WTdf, MTdf): \n",
    "    '''gibt abhängig davon, ob das Protein vom TF reguliert wird oder nicht, die Gleichung an, mit der der Flux (nur von MTdf???) berechnet wird'''\n",
    "\n",
    "    # if/else von WT oder MT?\n",
    "    \n",
    "    # linear          \n",
    "    if df[index,'TF_regulated'] == 0: #aus WT oder MT?\n",
    "        \n",
    "        return #was?\n",
    "    \n",
    "    # Hill\n",
    "    elif df[index,'TF_regulated'] == 'TF':\n",
    "\n",
    "        filt = (newdf['RctID']=='TF')\n",
    "        \n",
    "        Expr = MTdf.loc[index, 'Expression'].values\n",
    "        cTF = MTdf.loc[filt, 'Expression'].values # abhängig von geändertem TF-Promotor\n",
    "        K = WTdf.loc[index, 'Expression'].values # physiologischer Wert von c, Expression/Expr2Flux\n",
    "        n = 1.95 # from paper                \n",
    "        \n",
    "        Corr_ExprFlux = (2 * Expr) / (1+(K/cTF)**n)\n",
    "        \n",
    "        return #was?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Vergleich von WT und MT aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>1.441</td>\n",
       "      <td>GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG</td>\n",
       "      <td>ATGGGAACAAATTGA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>5.189023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.385</td>\n",
       "      <td>GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG</td>\n",
       "      <td>ATGCCAGTGTGCTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>1.123</td>\n",
       "      <td>GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG</td>\n",
       "      <td>ATGTTTATTGAATAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>4.328461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.813</td>\n",
       "      <td>GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG</td>\n",
       "      <td>ATGTATCAAATTTAA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-19.709134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.457</td>\n",
       "      <td>GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG</td>\n",
       "      <td>ATGATTAGCTTCTAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>3.404245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NADH16</td>\n",
       "      <td>2.394</td>\n",
       "      <td>GCCCATTGATAGCATATGTTACGGTACGTACTCTTGCACG</td>\n",
       "      <td>ATGACCGGTGTCTGA</td>\n",
       "      <td>38.534610</td>\n",
       "      <td>16.096328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NADTRHD</td>\n",
       "      <td>2.126</td>\n",
       "      <td>GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG</td>\n",
       "      <td>ATGTACAATGAATAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>2.358</td>\n",
       "      <td>GCCCATTGACATTGCTGTCTAGATAGTGTAGTTTTGCACG</td>\n",
       "      <td>ATGTGTTCGAACTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>2.020916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>3.341</td>\n",
       "      <td>GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG</td>\n",
       "      <td>ATGCCTCCCAGCTGA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>6.524841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>1.756</td>\n",
       "      <td>GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG</td>\n",
       "      <td>ATGGGTCGCGCTTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>5.286180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RctID  Expression                                  Promoter  \\\n",
       "0       PFK       1.441  GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG   \n",
       "1       PFL       1.385  GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG   \n",
       "2       PGI       1.123  GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG   \n",
       "3       PGK       0.813  GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG   \n",
       "4       PGL       1.457  GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG   \n",
       "..      ...         ...                                       ...   \n",
       "90   NADH16       2.394  GCCCATTGATAGCATATGTTACGGTACGTACTCTTGCACG   \n",
       "91  NADTRHD       2.126  GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG   \n",
       "92     NH4t       2.358  GCCCATTGACATTGCTGTCTAGATAGTGTAGTTTTGCACG   \n",
       "93      O2t       3.341  GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG   \n",
       "94      PDH       1.756  GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux  \n",
       "0   ATGGGAACAAATTGA   7.477382   5.189023  \n",
       "1   ATGCCAGTGTGCTAA   0.000000   0.000000  \n",
       "2   ATGTTTATTGAATAA   4.860861   4.328461  \n",
       "3   ATGTATCAAATTTAA -16.023526 -19.709134  \n",
       "4   ATGATTAGCTTCTAA   4.959985   3.404245  \n",
       "..              ...        ...        ...  \n",
       "90  ATGACCGGTGTCTGA  38.534610  16.096328  \n",
       "91  ATGTACAATGAATAA   0.000000   0.000000  \n",
       "92  ATGTGTTCGAACTAA   4.765319   2.020916  \n",
       "93  ATGCCTCCCAGCTGA  21.799493   6.524841  \n",
       "94  ATGGGTCGCGCTTAA   9.282533   5.286180  \n",
       "\n",
       "[95 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wtHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>1.441</td>\n",
       "      <td>GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG</td>\n",
       "      <td>ATGGGAACAAATTGA</td>\n",
       "      <td>7.438117</td>\n",
       "      <td>5.189023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.385</td>\n",
       "      <td>GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG</td>\n",
       "      <td>ATGCCAGTGTGCTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>1.123</td>\n",
       "      <td>GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG</td>\n",
       "      <td>ATGTTTATTGAATAA</td>\n",
       "      <td>4.804540</td>\n",
       "      <td>4.328461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.813</td>\n",
       "      <td>GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG</td>\n",
       "      <td>ATGTATCAAATTTAA</td>\n",
       "      <td>-15.954438</td>\n",
       "      <td>-19.709134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.457</td>\n",
       "      <td>GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG</td>\n",
       "      <td>ATGATTAGCTTCTAA</td>\n",
       "      <td>4.981618</td>\n",
       "      <td>3.404245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NADH16</td>\n",
       "      <td>2.394</td>\n",
       "      <td>GCCCATTGATAGCATATGTTACGGTACGTACTCTTGCACG</td>\n",
       "      <td>ATGACCGGTGTCTGA</td>\n",
       "      <td>38.429009</td>\n",
       "      <td>16.096328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NADTRHD</td>\n",
       "      <td>2.126</td>\n",
       "      <td>GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG</td>\n",
       "      <td>ATGTACAATGAATAA</td>\n",
       "      <td>0.073198</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>2.349</td>\n",
       "      <td>GCCCATTGACATTGCTGTCTAGATAGTGTTGTATTGCACG</td>\n",
       "      <td>ATGTGTTCGAACTAA</td>\n",
       "      <td>4.747131</td>\n",
       "      <td>2.020916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>3.341</td>\n",
       "      <td>GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG</td>\n",
       "      <td>ATGCCTCCCAGCTGA</td>\n",
       "      <td>21.733063</td>\n",
       "      <td>6.524841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>1.756</td>\n",
       "      <td>GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG</td>\n",
       "      <td>ATGGGTCGCGCTTAA</td>\n",
       "      <td>9.239173</td>\n",
       "      <td>5.286180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RctID  Expression                                  Promoter  \\\n",
       "0       PFK       1.441  GCCCATTGATAAGATAGTCAGCTCAAGGTAGTGTTGCACG   \n",
       "1       PFL       1.385  GCCCATTGATTCCCCACGAAGGCTTCTCTATTGTTGCACG   \n",
       "2       PGI       1.123  GCCCATTGATCGCCTCACGACGCCGCGTTAAGTTTGCACG   \n",
       "3       PGK       0.813  GCCCATTGATTTAAACGGATTACTAAAATAGGCTTGCACG   \n",
       "4       PGL       1.457  GCCCATTGAGTTAGGGCGGTGGGATGGGTATTTTTGCACG   \n",
       "..      ...         ...                                       ...   \n",
       "90   NADH16       2.394  GCCCATTGATAGCATATGTTACGGTACGTACTCTTGCACG   \n",
       "91  NADTRHD       2.126  GCCCATTGACTATGAACCCAGAAACCCCTAGCCTTGCACG   \n",
       "92     NH4t       2.349  GCCCATTGACATTGCTGTCTAGATAGTGTTGTATTGCACG   \n",
       "93      O2t       3.341  GCCCATTGACAGTCAACTCGGCTCTCTTTAAACTTGCACG   \n",
       "94      PDH       1.756  GCCCATTGAAGCTCAACTTGTGCCGGGCTATTCTTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux  \n",
       "0   ATGGGAACAAATTGA   7.438117   5.189023  \n",
       "1   ATGCCAGTGTGCTAA   0.000000   0.000000  \n",
       "2   ATGTTTATTGAATAA   4.804540   4.328461  \n",
       "3   ATGTATCAAATTTAA -15.954438 -19.709134  \n",
       "4   ATGATTAGCTTCTAA   4.981618   3.404245  \n",
       "..              ...        ...        ...  \n",
       "90  ATGACCGGTGTCTGA  38.429009  16.096328  \n",
       "91  ATGTACAATGAATAA   0.073198   0.000000  \n",
       "92  ATGTGTTCGAACTAA   4.747131   2.020916  \n",
       "93  ATGCCTCCCAGCTGA  21.733063   6.524841  \n",
       "94  ATGGGTCGCGCTTAA   9.239173   5.286180  \n",
       "\n",
       "[95 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression differences\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mylist = np.array(wtHost.strain.genes_df['Expression'])\n",
    "mylist2 = np.array(myHost.strain.genes_df['Expression'])\n",
    "print('Gene expression differences')\n",
    "np.round(mylist2/mylist,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exp2Flux diff\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-11-793ddc7a46d2>:5: RuntimeWarning: invalid value encountered in true_divide\n",
      "  np.round(myl2/myl,2)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 1., nan,  1.,  1.,  1., nan, nan,  1.,  1., nan, nan, nan,  1.,\n",
       "        1.,  1.,  1., nan, nan, nan, nan,  1.,  1., nan,  1.,  1., nan,\n",
       "        1.,  1.,  1.,  1., nan,  1., nan,  1., nan, nan,  1.,  1.,  1.,\n",
       "       nan,  1.,  1.,  1., nan, nan, nan,  1., nan, nan, nan, nan,  1.,\n",
       "       nan, nan,  1.,  1., nan, nan,  1.,  1.,  1., nan, nan,  1., nan,\n",
       "       nan, nan, nan, nan,  1., nan,  1.,  1.,  1.,  1., nan,  1., nan,\n",
       "       nan, nan,  1.,  1.,  1., nan, nan, nan, nan,  1., nan, nan,  1.,\n",
       "       nan,  1.,  1.,  1.])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myl = np.array(wtHost.strain.genes_df['Expr2Flux'])\n",
    "myl2 = np.array(myHost.strain.genes_df['Expr2Flux'])\n",
    "# myHost.strain.genes_df\n",
    "print('Exp2Flux diff')\n",
    "np.round(myl2/myl,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Flux reaction differences\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.99, 1.  , 0.99, 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ,\n",
       "       1.  , 1.  , 0.99, 0.99, 1.  , 1.  , 1.  , 1.  , 1.  , 0.99, 1.  ,\n",
       "       1.  , 0.99, 1.  , 1.  , 1.  , 1.01, 0.99, 1.  , 1.  , 1.  , 1.  ,\n",
       "       1.  , 1.  , 1.  , 0.99, 0.99, 1.01, 1.  , 1.01, 1.01, 0.99, 1.  ,\n",
       "       1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ,\n",
       "       1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 0.99, 1.  , 1.  ,\n",
       "       1.  , 1.  , 1.  , 0.99, 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ,\n",
       "       1.  , 1.  , 1.  , 1.  , 1.  , 0.99, 1.  , 1.  , 1.  , 1.  , 0.99,\n",
       "       1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FluxDiff = myHost.strain.genes_df['Fluxes']/wtHost.strain.genes_df['Fluxes']\n",
    "print('Flux reaction differences')\n",
    "np.round(np.nan_to_num(FluxDiff, nan=1, posinf=1, neginf=1),2)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "2001_QuantMiBi1_CobraPy_SimVal.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
