{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "6wxIUGOEReDX"
   },
   "outputs": [],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "from os.path import join\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# loading cobrapy, a library dedicated to the analysis of genome scale metabolic models\n",
    "# from cobra.io import read_sbml_model, write_sbml_model, load_matlab_model\n",
    "\n",
    "# loading escher for metabolic network visualization\n",
    "# import escher\n",
    "# from escher import Builder\n",
    "# from time import sleep\n",
    "# escher.rc['never_ask_before_quit'] = True\n",
    "# list of available maps\n",
    "# print(escher.list_available_maps())\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "# from memote import test_model, snapshot_report\n",
    "\n",
    "from biolabsim import Host, Strain, Ecol\n",
    "from biolabsim import measure_EnzymeLevel1, Help_GenomeGenerator, Strain, make_UpdateExpression, Help_StrainCharacterizer, Help_PromoterStrength, Help_Expr2Flux\n",
    "from Bio.Seq import Seq\n",
    "from pandas.core.frame import DataFrame\n",
    "from copy import deepcopy\n",
    "\n",
    "print('System ready')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ändern des Dataframes\n",
    "#### Hinzufügen des Transkriptionsfaktors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TranscriptionFactor(df:DataFrame, RegID:str, EnzymeTarget):\n",
    "    newdf = df\n",
    "    \n",
    "    # check if TF is already present, add columns to the default TF accordingly, no zero initiation of RegType column\n",
    "    if any('TF_regulated' in s for s in WTdf.columns.values):\n",
    "        TF = {\n",
    "        'RctID': [RegID], \n",
    "        'Expression': [2], # noch zu ändern mit Help_PromoterStrength()?\n",
    "        'Promoter':['GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG'],  # noch zu ändern (echt, random oder Fkt?)\n",
    "        'ORF': ['ATGGAGATGAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "        'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "        'Expr2Flux': [4.5], # noch zu ändern mit Help_Expr2Flux()?\n",
    "        'TF_regulated': [0],\n",
    "        'RegType': 'Hill_Activator'\n",
    "        }\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "        \n",
    "    else:\n",
    "    # add TF row:\n",
    "        Sequence = 'GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG'\n",
    "        TF = {\n",
    "        'RctID': [RegID], \n",
    "        'Expression': Help_PromoterStrength('Ecol', Sequence), # noch zu ändern mit Help_PromoterStrength()?\n",
    "        'Promoter': [Sequence],  # noch zu ändern (echt, random oder Fkt?)\n",
    "        'ORF': ['ATGCCCCCCAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "        'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "        'Expr2Flux': [4.5] # noch zu ändern mit Help_Expr2Flux()?\n",
    "        }\n",
    "\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "    \n",
    "        # add TF_regulated and RegType column:\n",
    "        zeroarray = np.zeros(len(newdf),dtype=int)\n",
    "        newdf = newdf.assign(TF_regulated=zeroarray) # später zu 0 oder jeweiliger TF-RctID ändern (random oder festgesetzt?)\n",
    "        newdf = newdf.assign(RegType=zeroarray)\n",
    "        \n",
    "        TF_Indx = (newdf['RctID']==RegID) # access TF row\n",
    "        newdf.loc[TF_Indx, 'RegType'] = 'Hill_Activator' # Metabolism Status: TF as activator\n",
    "    \n",
    "    # set Regulation of some Proteins:\n",
    "    for ETarget in EnzymeTarget:\n",
    "    # Future: fct recognizes what substrates are there, then makes TF_act or TF_rep\n",
    "        myIndx = (newdf['RctID']==ETarget)\n",
    "        newdf.loc[myIndx, 'TF_regulated'] = RegID\n",
    "\n",
    "    return newdf\n",
    "\n",
    "\n",
    "def make_DetectRegulatorPromoterMut(StrainWt:Strain, StrainMut:Strain, RctNewDF:DataFrame):\n",
    "#     for obj in (StrainWt, StrainMut):\n",
    "#         assert isinstance(obj, Strain), 'wrong strain type: must be Strain'\n",
    "#     assert isinstance(RctNewDF, DataFrame), 'ReactNewDF of wrong type: input dataframe'\n",
    "    GenesDF = StrainWt.genes_df\n",
    "    # finding regulators in GenesDF\n",
    "    myRegIndx = GenesDF[GenesDF['RegType'] != 0].index.values\n",
    "    # finding index of regulators with changes in the promoter\n",
    "    RegUpdatePromIndx = myRegIndx[RctNewDF.loc[myRegIndx, 'RctFlag'].values]\n",
    "    # finding name of regulators with changes in the promoter\n",
    "    RegUpdatePromName = RctNewDF.loc[RegUpdatePromIndx, 'RctID'].values\n",
    "    # finding enzymes regulated by regulator\n",
    "    EnzUpdateReguIndx = GenesDF['TF_regulated'].isin(RegUpdatePromName)\n",
    "    \n",
    "    return EnzUpdateReguIndx\n",
    "\n",
    "\n",
    "def measure_EnzymeLevel(HostName:str, StrainWT:Strain, StrainMut:Strain):\n",
    "    '''\n",
    "    The function differences of expression levels of enzymes between two strains.\n",
    "    '''\n",
    "    RefGenDF = StrainWT.genes_df\n",
    "    RefGenome = str(StrainWT.genome)\n",
    "    MutGenome = str(StrainMut.genome)\n",
    "    RefModel = StrainWT.model\n",
    "\n",
    "    RctNewDF = Help_StrainCharacterizer(HostName, RefGenDF, RefGenome, MutGenome, RefModel)\n",
    "    \n",
    "    # identifying changes in regulator expression\n",
    "    RegNewAr = make_DetectRegulatorPromoterMut(StrainWT, StrainMut, RctNewDF)\n",
    "\n",
    "    # combining promoter changes enzyme + regulator\n",
    "    AllNewAr = RctNewDF['RctFlag'].values | RegNewAr\n",
    "    AllNewDF = deepcopy(RctNewDF)\n",
    "    AllNewDF['RctFlag'] = AllNewAr  \n",
    "\n",
    "    return AllNewDF\n",
    "\n",
    "\n",
    "def calculate_StandardRegulatedExpression(WTdf):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf['TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "    print (RegTypes)\n",
    "\n",
    "    PreProcFlux = deepcopy(WTdf['Expression'].values)\n",
    "    \n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "        MaxExpr = 2 * WTdf.loc[EnzIndx, 'Expression'].values # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "        \n",
    "        cTF = WTdf.loc[RegIndx, 'Expression'].values # abhängig von geändertem TF-Promotor\n",
    "        K = WTdf.loc[EnzIndx, 'Expression'].values # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95 # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.  \n",
    "#         print ('Eqn: ', RegType, MaxExpr, cTF, K)\n",
    "        \n",
    "        if EquType == 'Hill_Activator':\n",
    "            Flux = MaxExpr / (1+(K/cTF)**n)\n",
    "#             print(Flux)\n",
    "   \n",
    "        elif EquType == 'Hill_Repressor':\n",
    "            Flux = MaxExpr * (1 - 1/ (1+(K/cTF)**n))\n",
    "\n",
    "#         print(EnzIndx)\n",
    "        PreProcFlux[EnzIndx] = Flux\n",
    "\n",
    "    return PreProcFlux\n",
    "\n",
    "    \n",
    "def calculate_mutatedRegulatedExpression(WTdf, AllNewDF):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf.loc[AllNewDF['RctFlag'], 'TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "\n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "        MaxExpr = 2 * AllNewDF.loc[EnzIndx, 'RefExpr'].values # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "\n",
    "        cTF = AllNewDF.loc[RegIndx, 'NewExpr'].values # abhängig von geändertem TF-Promotor\n",
    "        K = AllNewDF.loc[EnzIndx, 'RefExpr'].values # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95 # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.  \n",
    "#         print ('Eqn: ', AllNewDF)\n",
    "        \n",
    "        if EquType == 'Hill_Activator':\n",
    "            Flux = MaxExpr / (1+(K/cTF)**n)\n",
    "#             print(Flux)\n",
    "   \n",
    "        elif EquType == 'Hill_Repressor':\n",
    "            Flux = MaxExpr * (1 - 1/ (1+(K/cTF)**n))\n",
    "\n",
    "            \n",
    "        PreProcFlux = AllNewDF['NewExpr'].values\n",
    "        PreProcFlux[EnzIndx] = Flux\n",
    "\n",
    "    return PreProcFlux\n",
    "        \n",
    "def setboundary(WTdf, AllNewDF):\n",
    "    '''\n",
    "    Finding and setting the boundaries for cobrapy.\n",
    "    '''\n",
    "\n",
    "    # finding reactions for which the expression has changed\n",
    "    OnlyEnzIndx = WTdf['RegType'] == 0\n",
    "    RctNewDF = AllNewDF.loc[OnlyEnzIndx, ['RctFlag']] == True\n",
    "    RctNew = np.arange((len(RctNewDF)))[np.ravel(RctNewDF.values)]\n",
    "#     OnlyEnzIndx = WTdf[WTdf['RegType'] == 0].index.values\n",
    "#     RctNew = AllNewDF[AllNewDF.loc[OnlyEnzIndx, 'RctFlag'] == True].index.values\n",
    "    \n",
    "    # For reactions with reduced and positive flux: reduce the upper limit,\n",
    "    # For reactions with increased and positive flux: increase the lower limit\n",
    "    # For reactions with negative flux the limits are exchanged.\n",
    "    FluxPos = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux']>0])]\n",
    "    FluxNeg = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux']<0])]\n",
    "    # Finding increased and decreased fluxes\n",
    "    FluxInc = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values>1]\n",
    "    FluxDec = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values<1]\n",
    "\n",
    "    # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "    PosIncInd = np.intersect1d(FluxPos,FluxInc)\n",
    "    # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "    PosDecInd = np.intersect1d(FluxPos,FluxDec)\n",
    "    # Comb.3: negative flux with increased expression -> decreasing lower bound\n",
    "    NegIncInd = np.intersect1d(FluxNeg,FluxInc)\n",
    "    # Comb.4: positive flux with increased expression -> increasing lower bound\n",
    "    NegDecInd = np.intersect1d(FluxNeg,FluxDec)\n",
    "\n",
    "    Expr_Change = {'increase': np.hstack([PosIncInd,NegIncInd]),'decrease': np.hstack([PosDecInd,NegDecInd])}\n",
    "    Set_Boundary = {'lower': np.hstack([PosIncInd,NegDecInd]),'upper': np.hstack([PosDecInd,NegIncInd])}\n",
    "    \n",
    "#     Set_Boundary = {'lower': [0,2,3],'upper': []}\n",
    "    return Set_Boundary\n",
    "\n",
    "# Finding reactions to change. Three possibilities exits: 1. the promoter of the enzyme itself has changed, 2. the promoter of a regulator has changed, 3. the enzyme promoter and the regulator promoter have changed\n",
    "# redefining Help_FluxCalculator in simulation/metabolism\n",
    "def Help_FluxCalculator ( HostName:str, Strain:Strain, AllNewDF = None ) :\n",
    "    '''\n",
    "    Calculation of flux values.\n",
    "\n",
    "    This method can work in 2 modes:\n",
    "      [Strain only] : The metabolic model of the WT strain is used for calculation.\n",
    "      [Strain + AllNewDF] : An additional step of resetting boundaries is done on the model\n",
    "        before the fluxes are calculated.\n",
    "\n",
    "    TODO: The \"reset boundary\" step is mutating the `StrainMut.model`, mutation might not be intended.\n",
    "      Because of this, perhaps the other `Strain.model` gets mutated in the process.\n",
    "    '''\n",
    "#     from ..measurement.fluxes import measure_EnzymeLevel1\n",
    "\n",
    "    # adding flux values\n",
    "    # setup of flux boundaries. For the reference boundary changes are set to 'False',\n",
    "    # for mutant strains, ractions with altered promoter sequence will change enzyme levels and boundaries must be changed accordingly, their variable is 'True'\n",
    "\n",
    "    if AllNewDF is not None :\n",
    "        print('resetting boundaries')\n",
    "        # finding reactions for which the expression has changed, and finding the new level\n",
    "        RefFlux = AllNewDF['RefFlux']\n",
    "        NewFlux = AllNewDF['NewFlux']\n",
    "        \n",
    "        Set_Boundary = setboundary(Strain.genes_df, AllNewDF)\n",
    "        # Defining the model with the two combinations of either\n",
    "        # increasing lower bound (increased forward, decreased reverse reaction)\n",
    "        # decreasing upper bound (decreased forward, increased reverse reaction)\n",
    "        with Strain.model as myModel:\n",
    "            # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "            for Indx in Set_Boundary['lower']:\n",
    "                myModel.reactions[Indx].lower_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "            # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "            for Indx in Set_Boundary['upper']:\n",
    "                myModel.reactions[Indx].upper_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "\n",
    "            Fluxes = myModel.optimize()\n",
    "\n",
    "    else:\n",
    "        Fluxes = Strain.model.optimize()\n",
    "\n",
    "\n",
    "    return Fluxes.fluxes.values, Fluxes.objective_value\n",
    "\n",
    "\n",
    "print('Functions defined.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wtHost = Ecol()\n",
    "print('Wild type growth rate: {:.2f}'.format(wtHost.strain.objective))\n",
    "\n",
    "# adding regulator TF to dataframe and genome of wt\n",
    "WTdf = wtHost.strain.genes_df\n",
    "WTdf = add_TranscriptionFactor(WTdf, 'TF1', ['PFK','PGI'])\n",
    "WTdf = add_TranscriptionFactor(WTdf, 'TF2', ['PGL'])\n",
    "# update for regulated expression\n",
    "corrExpr = calculate_StandardRegulatedExpression(WTdf)\n",
    "WTdf['Expression'] = corrExpr\n",
    "# update for Expression to Flux correlation, for all a linear factor\n",
    "myExpr2Flux = Help_Expr2Flux(WTdf['Fluxes'], corrExpr)\n",
    "WTdf['Expr2Flux'] = myExpr2Flux\n",
    "# putting the genes dataframe back to the wild type\n",
    "wtHost.strain.genes_df = WTdf\n",
    "# Generating new Genome\n",
    "WTGenome_new = Help_GenomeGenerator(WTdf, 500, .6)\n",
    "wtHost.strain.genome = Seq(WTGenome_new)\n",
    "\n",
    "wtHost.strain.genes_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generating mutant\n",
    "# We will actually define the mutations by hand and delete the automatically generated genome for now\n",
    "myHost = deepcopy(wtHost)\n",
    "# Mutants to be generated\n",
    "Mutant_Targets = ['PFK','PGK','TF1']\n",
    "\n",
    "for myMutant in Mutant_Targets:\n",
    "#     Mutant_Bool = myHost.strain.genes_df['RctID']==myMutant\n",
    "#     Mutant_Indx = tuple(np.arange(len(Mutant_Bool))[Mutant_Bool])\n",
    "    MTdf = myHost.strain.genes_df.copy()\n",
    "    Mutant_Indx = tuple(MTdf[MTdf['RctID'] == myMutant].index.values)\n",
    "    # Finding the index for later printing\n",
    "    WTIndx = str(wtHost.strain.genome).find(wtHost.strain.genes_df.loc[Mutant_Indx, 'Promoter'])\n",
    "#     # Generating target promoter sequence, ATTGA is always there\n",
    "    PromTar = MTdf.loc[Mutant_Indx, 'Promoter'].replace('ATTGA','CCCCC')\n",
    "    # converting Biopython Seq class to string. This makes string replacements easier\n",
    "    MutGenome = str(myHost.strain.genome)\n",
    "    MutGenome = MutGenome.replace(MTdf.loc[Mutant_Indx, 'Promoter'], PromTar)\n",
    "    myHost.strain.genome = Seq(MutGenome)\n",
    "    MTdf.loc[Mutant_Indx, 'Promoter'] = PromTar\n",
    "    myHost.strain.genes_df = MTdf.copy()\n",
    "#     print('{}, Reference:\\t{},\\nMutated:\\t{}\\nwith index {}'.format(myMutant, wtHost.strain.genome[Mutant_Indx:Mutant_Indx+41], myHost.strain.genome[Mutant_Indx:Mutant_Indx+41], Mutant_Indx))\n",
    "    print(myMutant, WTIndx)\n",
    "\n",
    "# updating expression of regulated enzymes\n",
    "AllNewDF = measure_EnzymeLevel('Ecol', wtHost.strain, myHost.strain)\n",
    "CorrRegExpr = calculate_mutatedRegulatedExpression(WTdf, AllNewDF)\n",
    "AllNewDF['NewExpr'] = CorrRegExpr\n",
    "AllNewDF['NewFlux'] = AllNewDF['NewExpr']*WTdf['Expr2Flux']\n",
    "AllNewDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "solution,_ = Help_FluxCalculator('Ecol', myHost.strain, AllNewDF)\n",
    "RefFluxDict = WTdf.set_index('RctID')['Fluxes'].to_dict()\n",
    "NewFluxDict = {a:b for a,b in zip(WTdf['RctID'].values, solution)}\n",
    "AllFluxDict = [RefFluxDict, NewFluxDict]\n",
    "# {WTdf['RctID'].tolist(): solution}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Vergleich von WT und MT aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import escher\n",
    "from escher import Builder\n",
    "from time import sleep\n",
    "escher.rc['never_ask_before_quit'] = True\n",
    "# list of available maps\n",
    "print(escher.list_available_maps())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "-wAeRsXJQI6e"
   },
   "outputs": [],
   "source": [
    "builder = Builder()\n",
    "# Load an Escher map\n",
    "builder.map_name = 'iJO1366.Central metabolism'\n",
    "builder.model = wtHost.strain.model\n",
    "builder.reaction_data = AllFluxDict\n",
    "# Add some data for metabolites\n",
    "#builder.metabolite_data = solution.shadow_prices\n",
    "# Simplify the map by hiding some labels\n",
    "builder.hide_secondary_metabolites = True\n",
    "#builder.hide_all_labels = True\n",
    "builder.reaction_scale = [\n",
    "    { 'type': 'min', 'color': '#000000', 'size': 12 },\n",
    "    { 'type': 'median', 'color': '#ffffff', 'size': 20 },\n",
    "    { 'type': 'max', 'color': '#ff0000', 'size': 25 }\n",
    "]\n",
    "builder.reaction_scale_preset = 'GaBuRd'\n",
    "\n",
    "# Make all the arrows three times as thick\n",
    "builder.reaction_scale = [\n",
    "    {k: v * 3 if k == 'size' else v for k, v in x.items()}\n",
    "    for x in builder.reaction_scale\n",
    "]\n",
    "builder.save_html('Escher-Metabol-Map.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wtHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mylist = np.array(wtHost.strain.genes_df['Expression'])\n",
    "mylist2 = np.array(myHost.strain.genes_df['Expression'])\n",
    "print('Gene expression differences')\n",
    "np.round(mylist2/mylist,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myl = np.array(wtHost.strain.genes_df['Expr2Flux'])\n",
    "myl2 = np.array(myHost.strain.genes_df['Expr2Flux'])\n",
    "# myHost.strain.genes_df\n",
    "print('Exp2Flux diff')\n",
    "np.round(myl2/myl,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FluxDiff = myHost.strain.genes_df['Fluxes']/wtHost.strain.genes_df['Fluxes']\n",
    "print('Flux reaction differences')\n",
    "np.round(np.nan_to_num(FluxDiff, nan=1, posinf=1, neginf=1),2)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "2001_QuantMiBi1_CobraPy_SimVal.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
