{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "6wxIUGOEReDX"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System ready\n"
     ]
    }
   ],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "from os.path import join\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# loading cobrapy, a library dedicated to the analysis of genome scale metabolic models\n",
    "# from cobra.io import read_sbml_model, write_sbml_model, load_matlab_model\n",
    "\n",
    "# loading escher for metabolic network visualization\n",
    "# import escher\n",
    "# from escher import Builder\n",
    "# from time import sleep\n",
    "# escher.rc['never_ask_before_quit'] = True\n",
    "# list of available maps\n",
    "# print(escher.list_available_maps())\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "# from memote import test_model, snapshot_report\n",
    "\n",
    "from biolabsim import Host, Strain, Ecol\n",
    "from biolabsim import measure_EnzymeLevel1, Help_GenomeGenerator, Help_PromoterStrength\n",
    "from Bio.Seq import Seq\n",
    "from copy import deepcopy\n",
    "\n",
    "print('System ready')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Change Setting\n",
    "#### Add Transkriptionsfaktor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TranscriptionFactor(df, TFname1:str, TFname2:str):\n",
    "    newdf = df\n",
    "    Promoter1 = 'GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG' # noch zu ändern? (echt, random oder Fkt?)\n",
    "    # add TF rows:\n",
    "    TF1 = {\n",
    "    'RctID':[TFname1], \n",
    "    #(Host, Sequence, Scaler=1, Similarity_Thresh=.4, Predict_File=None):\n",
    "    'Expression': [Help_PromoterStrength('Ecol',Promoter1)],\n",
    "    'Promoter': [Promoter1],  \n",
    "    'ORF': ['ATGGAGATGAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "    'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "    'Expr2Flux': [4.5] # noch zu ändern mit Help_Expr2Flux()?\n",
    "    }\n",
    "    dfTF1 = pd.DataFrame(TF1)\n",
    "    newdf = newdf.append(dfTF1, ignore_index=True)\n",
    "    \n",
    "    Promoter2 = 'GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCATT'\n",
    "    TF2 = {\n",
    "    'RctID':[TFname2], \n",
    "    'Expression':[Help_PromoterStrength('Ecol',Promoter2)], # noch zu ändern mit Help_PromoterStrength()?\n",
    "    'Promoter':[Promoter2],  # noch zu ändern (echt, random oder Fkt?)\n",
    "    'ORF': ['ATGGAGATGAAGTCC'], # noch zu ändern (echt, random oder Fkt?)\n",
    "    'Fluxes': [6], # noch zu ändern mit Help_FluxCalculator()?\n",
    "    'Expr2Flux': [2] # noch zu ändern mit Help_Expr2Flux()?\n",
    "    }\n",
    "    dfTF2 = pd.DataFrame(TF2)\n",
    "    newdf = newdf.append(dfTF2, ignore_index=True)\n",
    "    \n",
    "    # add TF_regulated and RegType column:\n",
    "    zeroarray = np.zeros(len(newdf),dtype=int)\n",
    "    newdf = newdf.assign(TF_regulated=zeroarray) # später zu 0 oder jeweiliger TF-RctID ändern (random oder festgesetzt?)\n",
    "    newdf = newdf.assign(RegType=zeroarray)\n",
    "    \n",
    "    # set Regulation of some Proteins:\n",
    "    TFname1_Indx = (newdf['RctID']==TFname1) # access TF1 row\n",
    "    newdf.loc[TFname1_Indx, 'RegType'] = 'Hill_Activator' # Metabolism Status: TF as activator\n",
    "    TFname2_Indx = (newdf['RctID']==TFname2) # access TF1 row\n",
    "    newdf.loc[TFname2_Indx, 'RegType'] = 'Hill_Repressor' # Metabolism Status: TF as repressor\n",
    "    # Future: fct recognizes what substrates are there, then makes TF_act or TF_rep\n",
    "    PFK_Indx = (newdf['RctID']=='PFK')\n",
    "    newdf.loc[PFK_Indx, 'TF_regulated'] = TFname1\n",
    "    PGI_Indx = (newdf['RctID']=='PGI')\n",
    "    newdf.loc[PGI_Indx, 'TF_regulated'] = TFname1 \n",
    "    PGL_Indx = (newdf['RctID']=='PGL')\n",
    "    newdf.loc[PGL_Indx, 'TF_regulated'] = TFname2\n",
    "    return newdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Change Wildtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wild type growth rate: 0.87\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>TF_regulated</th>\n",
       "      <th>RegType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>1.832</td>\n",
       "      <td>GCCCATTGATTCCGACGGCGGGCACTGGTACAATTGCACG</td>\n",
       "      <td>ATGACCGGCTTGTGA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>4.081540</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>3.470</td>\n",
       "      <td>GCCCATTGACCTATCGCCCCTCGCCCTATACAATTGCACG</td>\n",
       "      <td>ATGGACTTTAAGTGA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>0.438</td>\n",
       "      <td>GCCCATTGAGTACGTAGTGCGCACGACGTAGGGTTGCACG</td>\n",
       "      <td>ATGCTGGTAGGATAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>11.097856</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>1.134</td>\n",
       "      <td>GCCCATTGAGCATCATGTACCACGCCTGTAGTGTTGCACG</td>\n",
       "      <td>ATGCCCAATGAATGA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-14.130094</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.226</td>\n",
       "      <td>GCCCATTGAAGGAGGTGAAGGGTTAGAATAGTCTTGCACG</td>\n",
       "      <td>ATGGGCGATTCATAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>4.045665</td>\n",
       "      <td>TF2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.835</td>\n",
       "      <td>GCCCATTGACCGCGGGAGGTAAATGTTCTAAGATTGCACG</td>\n",
       "      <td>ATGTTCCAGCTGTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>2.596904</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>0.991</td>\n",
       "      <td>GCCCATTGAGCCGCGCGGTTATTGGAAATACTTTTGCACG</td>\n",
       "      <td>ATGATTAATACGTGA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>21.997470</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>0.735</td>\n",
       "      <td>GCCCATTGATATCCGTCCGTCTCTCCAATAGCCTTGCACG</td>\n",
       "      <td>ATGCCTCTAATGTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>12.629296</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG</td>\n",
       "      <td>ATGGAGATGAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TF2</td>\n",
       "      <td>5.537</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCATT</td>\n",
       "      <td>ATGGAGATGAAGTCC</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Repressor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   RctID  Expression                                  Promoter  \\\n",
       "0    PFK       1.832  GCCCATTGATTCCGACGGCGGGCACTGGTACAATTGCACG   \n",
       "1    PFL       3.470  GCCCATTGACCTATCGCCCCTCGCCCTATACAATTGCACG   \n",
       "2    PGI       0.438  GCCCATTGAGTACGTAGTGCGCACGACGTAGGGTTGCACG   \n",
       "3    PGK       1.134  GCCCATTGAGCATCATGTACCACGCCTGTAGTGTTGCACG   \n",
       "4    PGL       1.226  GCCCATTGAAGGAGGTGAAGGGTTAGAATAGTCTTGCACG   \n",
       "..   ...         ...                                       ...   \n",
       "92  NH4t       1.835  GCCCATTGACCGCGGGAGGTAAATGTTCTAAGATTGCACG   \n",
       "93   O2t       0.991  GCCCATTGAGCCGCGCGGTTATTGGAAATACTTTTGCACG   \n",
       "94   PDH       0.735  GCCCATTGATATCCGTCCGTCTCTCCAATAGCCTTGCACG   \n",
       "95   TF1       5.733  GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG   \n",
       "96   TF2       5.537  GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCATT   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux TF_regulated         RegType  \n",
       "0   ATGACCGGCTTGTGA   7.477382   4.081540          TF1               0  \n",
       "1   ATGGACTTTAAGTGA   0.000000   0.000000            0               0  \n",
       "2   ATGCTGGTAGGATAA   4.860861  11.097856          TF1               0  \n",
       "3   ATGCCCAATGAATGA -16.023526 -14.130094            0               0  \n",
       "4   ATGGGCGATTCATAA   4.959985   4.045665          TF2               0  \n",
       "..              ...        ...        ...          ...             ...  \n",
       "92  ATGTTCCAGCTGTAA   4.765319   2.596904            0               0  \n",
       "93  ATGATTAATACGTGA  21.799493  21.997470            0               0  \n",
       "94  ATGCCTCTAATGTAA   9.282533  12.629296            0               0  \n",
       "95  ATGGAGATGAAGTAA   9.000000   4.500000            0  Hill_Activator  \n",
       "96  ATGGAGATGAAGTCC   6.000000   2.000000            0  Hill_Repressor  \n",
       "\n",
       "[97 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wtHost = Ecol()\n",
    "print('Wild type growth rate: {:.2f}'.format(wtHost.strain.objective))\n",
    "\n",
    "# adding regulator TF to dataframe and genome of wt\n",
    "WTdf = wtHost.strain.genes_df\n",
    "WTdf = add_TranscriptionFactor(WTdf, 'TF1', 'TF2')\n",
    "wtHost.strain.genes_df = WTdf\n",
    "# Generating new Genome\n",
    "WTGenome_new = Help_GenomeGenerator(WTdf, 500, .6)\n",
    "wtHost.strain.genome = Seq(WTGenome_new)\n",
    "wtHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Change Mutant\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PFK 0\n",
      "PGK 183\n",
      "PGL 239\n",
      "TF1 5718\n"
     ]
    }
   ],
   "source": [
    "# Generating mutant\n",
    "# We will actually define the mutations by hand and delete the automatically generated genome for now\n",
    "myHost = deepcopy(wtHost)\n",
    "# Mutants to be generated\n",
    "Mutant_Targets = ['PFK','PGK','PGL','TF1']\n",
    "\n",
    "for myMutant in Mutant_Targets:\n",
    "    Mutant_Bool = myHost.strain.genes_df['RctID']==myMutant\n",
    "    Mutant_Indx = tuple(np.arange(len(Mutant_Bool))[Mutant_Bool])\n",
    "    MTdf = myHost.strain.genes_df.copy()\n",
    "    # Finding the index for later printing\n",
    "    WTIndx = str(wtHost.strain.genome).find(wtHost.strain.genes_df.loc[Mutant_Indx, 'Promoter'])\n",
    "#     # Generating target promoter sequence, ATTGA is always there\n",
    "    PromTar = MTdf.loc[Mutant_Indx, 'Promoter'].replace('ATTGA','CCCCC')\n",
    "    # converting Biopython Seq class to string. This makes string replacements easier\n",
    "    MutGenome = str(myHost.strain.genome)\n",
    "    MutGenome = MutGenome.replace(MTdf.loc[Mutant_Indx, 'Promoter'], PromTar)\n",
    "    myHost.strain.genome = Seq(MutGenome)\n",
    "#     print('{}, Reference:\\t{},\\nMutated:\\t{}\\nwith index {}'.format(myMutant, wtHost.strain.genome[Mutant_Indx:Mutant_Indx+41], myHost.strain.genome[Mutant_Indx:Mutant_Indx+41], Mutant_Indx))\n",
    "    MTdf.loc[Mutant_Indx, 'Promoter'] = PromTar\n",
    "    myHost.strain.genes_df = MTdf\n",
    "    print(myMutant, WTIndx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing measure_EnzymeLevel1\n",
    "qw,we,er = measure_EnzymeLevel1('Ecol', wtHost.strain, myHost.strain)\n",
    "MT_Filt = qw['RctFlag']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctFlag</th>\n",
       "      <th>RctID</th>\n",
       "      <th>RefExpr</th>\n",
       "      <th>NewExpr</th>\n",
       "      <th>RefFlux</th>\n",
       "      <th>Expr2Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>PFK</td>\n",
       "      <td>1.832</td>\n",
       "      <td>1.832</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>4.081540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>PFL</td>\n",
       "      <td>3.470</td>\n",
       "      <td>3.470</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>PGI</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.438</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>11.097856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>PGK</td>\n",
       "      <td>1.134</td>\n",
       "      <td>1.134</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-14.130094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>PGL</td>\n",
       "      <td>1.226</td>\n",
       "      <td>1.226</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>4.045665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>False</td>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.835</td>\n",
       "      <td>1.835</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>2.596904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>False</td>\n",
       "      <td>O2t</td>\n",
       "      <td>0.991</td>\n",
       "      <td>0.991</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>21.997470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>False</td>\n",
       "      <td>PDH</td>\n",
       "      <td>0.735</td>\n",
       "      <td>0.735</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>12.629296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>True</td>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733</td>\n",
       "      <td>5.733</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>False</td>\n",
       "      <td>TF2</td>\n",
       "      <td>5.537</td>\n",
       "      <td>5.537</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    RctFlag RctID  RefExpr  NewExpr    RefFlux  Expr2Flux\n",
       "0      True   PFK    1.832    1.832   7.477382   4.081540\n",
       "1     False   PFL    3.470    3.470   0.000000   0.000000\n",
       "2     False   PGI    0.438    0.438   4.860861  11.097856\n",
       "3      True   PGK    1.134    1.134 -16.023526 -14.130094\n",
       "4      True   PGL    1.226    1.226   4.959985   4.045665\n",
       "..      ...   ...      ...      ...        ...        ...\n",
       "92    False  NH4t    1.835    1.835   4.765319   2.596904\n",
       "93    False   O2t    0.991    0.991  21.799493  21.997470\n",
       "94    False   PDH    0.735    0.735   9.282533  12.629296\n",
       "95     True   TF1    5.733    5.733   9.000000   4.500000\n",
       "96    False   TF2    5.537    5.537   6.000000   2.000000\n",
       "\n",
       "[97 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qw # = RctNewDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "we # = EnzymeProm_Mutated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "er "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MTdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# measure_RegulatorLevel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def measure_RegulatorLevel(HostName:str, StrainWT:Strain, StrainMut:Strain, MT_Filt):\n",
    "    'find enzymes regulated by mutated TFs'\n",
    "    from biolabsim.simulation.metabolism import Help_StrainCharacterizer\n",
    "\n",
    "    RefGenDF = StrainWT.genes_df\n",
    "    RefGenome = str(StrainWT.genome)\n",
    "    MutGenome = str(StrainMut.genome)\n",
    "    RefModel = StrainWT.model\n",
    "\n",
    "    # Changes from measure_EnzymeLevel1\n",
    "    TF_Indx = (RefGenDF['RegType'] != 0) # filter for TFs\n",
    "    MT_TF = TF_Indx & MT_Filt # filter for TFs with mutations \n",
    "    MT_TFnames = RefGenDF.loc[MT_TF, 'RctID'].values # names of mutated TFs\n",
    "    # MT_TF_indx = np.where(MT_TF == True)[0] # indexes of mutated TFs\n",
    "    MT_TFenzymes = RefGenDF.TF_regulated.isin(MT_TFnames) # filter for enzymes regulated by mutated TFs\n",
    "    alteredEnzymes = MT_Filt | MT_TFenzymes # filter for Enzymes that are mutated or regulated by mutated TFs\n",
    "    TFNewDF = Help_StrainCharacterizer(HostName, RefGenDF, RefGenome, MutGenome, RefModel)\n",
    "    TFNewDF['RctFlag'] = alteredEnzymes # convert first column from reactions with changed expression to enzymes for which TF is mutated\n",
    "    RctNewDF = TFNewDF\n",
    "    \n",
    "#     #berechnen für alle statt bestimmte\n",
    "#     RctNew = RctNewDF[RctNewDF['RctFlag']==True].index.values\n",
    "#     # For reactions with reduced expression and positive flux: reduce the upper limit,\n",
    "#     # For reactions with increased expression and positive flux: increase the lower limit\n",
    "#     # for reactions with negative flux the limits are exchanged.\n",
    "#     FluxPos = RctNew[tuple([RctNewDF.loc[RctNew, 'RefFlux']>0])]\n",
    "#     FluxNeg = RctNew[tuple([RctNewDF.loc[RctNew, 'RefFlux']<0])]\n",
    "#     # Finding increased and decreased fluxes\n",
    "#     FluxInc = RctNew[RctNewDF.loc[RctNew, 'NewExpr'].values / RctNewDF.loc[RctNew, 'RefExpr'].values>1]\n",
    "#     FluxEqu = RctNew[RctNewDF.loc[RctNew, 'NewExpr'].values / RctNewDF.loc[RctNew, 'RefExpr'].values=1]\n",
    "#     FluxDec = RctNew[RctNewDF.loc[RctNew, 'NewExpr'].values / RctNewDF.loc[RctNew, 'RefExpr'].values<1]\n",
    "\n",
    "#     # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "#     PosIncInd = np.intersect1d(FluxPos,FluxInc)\n",
    "#     # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "#     PosDecInd = np.intersect1d(FluxPos,FluxDec)\n",
    "#     # Comb.3: negative flux with increased expression -> decreasing lower bound\n",
    "#     NegIncInd = np.intersect1d(FluxNeg,FluxInc)\n",
    "#     # Comb.4: positive flux with increased expression -> increasing lower bound\n",
    "#     NegDecInd = np.intersect1d(FluxNeg,FluxDec)\n",
    "\n",
    "#     Expr_Change = {'increase': np.hstack([PosIncInd,NegIncInd]),'decrease': np.hstack([PosDecInd,NegDecInd])}\n",
    "#     Set_Boundary = {'lower': np.hstack([PosIncInd,NegDecInd]),'upper': np.hstack([PosDecInd,NegIncInd])}\n",
    "\n",
    "    return RctNewDF # , Set_Boundary, Expr_Change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing measure_RegulatorLevel\n",
    "rt= measure_RegulatorLevel('Ecol', wtHost.strain, myHost.strain, MT_Filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# try out stuff for measure_RegulatorLevel()\n",
    "# GenesDF=MTdf\n",
    "# TF_Indx = (GenesDF['RegType'] != 0) # Filter for TFs\n",
    "# MT_TF = TF_Indx & MT_Filt # TFs with mutations \n",
    "    \n",
    "# MT_TFnames = GenesDF.loc[MT_TF, 'RctID'].values # names of mutated TFs (array)\n",
    "# MT_TFenzymes = GenesDF.TF_regulated.isin(MT_TFnames) # Enzymes regulated by mt TFs # RctNewDF =\n",
    "\n",
    "# print(MT_TFnames)\n",
    "# MT_TFenzymes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regulationsgleichung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# used in Help_FluxCalculator()\n",
    "\n",
    "def FluxEquation(index, WTdf, TFNewDF): \n",
    "    '''gibt abhängig davon, ob das Protein vom TF reguliert wird oder nicht, die Gleichung an, mit der der Flux berechnet wird'''\n",
    "          \n",
    "    myTF = WTdf.loc[index, 'TF_regulated']\n",
    "    \n",
    "    \n",
    "    if WTdf[index,'TF_regulated'] == 0: \n",
    "        # linear equation\n",
    "        \n",
    "        Corr_ExprFlux = TFNewDF.loc[index, 'Expr2Flux'] # aus measureenzymelevel oder measureRegulator?\n",
    "        Flux = TFNewDF.loc[index, 'NewExpr'] * Corr_ExprFlux \n",
    "        return Flux\n",
    "    \n",
    "    else:\n",
    "        # Hill-Equation (from Kim, O'Shea (2008): A quantitative model of transcription factor–activated gene expression)\n",
    "        \n",
    "        RegIndx = (WTdf['RctID']==WTdf.loc[index,'TF_regulated']) # filter for the corresponding TF\n",
    "        \n",
    "        MaxExpr = 2 * TFNewDF.loc[index, 'NewExpr'].values # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "        cTF = TFNewDF.loc[RegIndx, 'NewExpr'].values # abhängig von geändertem TF-Promotor\n",
    "        K = TFNewDF.loc[index, 'RefExpr'].values # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95 # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.  \n",
    "        \n",
    "        \n",
    "        if WTdf[RegIndx,'RegType'] == 'Hill_Activator':\n",
    "\n",
    "            Flux = MaxExpr / (1+(K/cTF)**n)\n",
    "\n",
    "    \n",
    "        elif WTdf[RegIndx,'RegType'] == 'Hill_Repressor':\n",
    "\n",
    "            Flux = MaxExpr * (1 - 1/ (1+(K/cTF)**n))\n",
    "\n",
    "        return Flux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setboundary(RctNewDF, StrainWT:Strain):\n",
    "    'Dict of lower and upper changes'\n",
    "    lower =  \n",
    "    upper = [0, 2]\n",
    "    Set_Boundary = {'lower': lower,\n",
    "                   'upper': upper}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding reactions to change. Three possibilities exits: 1. the promoter of the enzyme itself has changed, 2. the promoter of a regulator has changed, \n",
    "# 3. the enzyme promoter and the regulator promoter have changed --> found with rt['RctFLag'] from measure_RegulatorLevel()\n",
    "# redefining Help_FluxCalculator in simulation/metabolism\n",
    "def Help_FluxCalculator ( HostName:str, StrainWT:Strain, StrainMut = None ) :\n",
    "    '''\n",
    "    Calculation of flux values.\n",
    "\n",
    "    This method can work in 2 modes:\n",
    "      [StrainWT only] : The metabolic model of the WT strain is used for calculation.\n",
    "      [StrainWT + StrainMut] : An additional step of resetting boundaries is done on the model\n",
    "        before the fluxes are calculated.\n",
    "\n",
    "    TODO: The \"reset boundary\" step is mutating the `StrainMut.model`, mutation might not be intended.\n",
    "      Because of this, perhaps the other `StrainWT.model` gets mutated in the process.\n",
    "    '''\n",
    "    from ..measurement.fluxes import measure_EnzymeLevel1\n",
    "\n",
    "    # adding flux values\n",
    "    # setup of flux boundaries. For the reference boundary changes are set to 'False',\n",
    "    # for mutant strains, ractions with altered promoter sequence will change enzyme levels and boundaries must be changed accordingly, their variable is 'True'\n",
    "\n",
    "    if StrainMut is not None :\n",
    "        print('resetting boundaries')\n",
    "        # finding reactions for which the expression has changed, and finding the new level\n",
    "        RctNewDF, EnzymeProm_Mutated, _ = measure_EnzymeLevel1(HostName, StrainWT, StrainMut)\n",
    "        # finding regulators whose promoter has changed, and finding the new level\n",
    "        MT_Filt = RctNewDF['RctFlag']\n",
    "        AllNewDF = measure_RegulatorLevel(HostName, StrainWT, StrainMut, MT_Filt)  # ???\n",
    "        # hier wird doch dann RctNewDF überscrhieben, soll das so?\n",
    "        # oder nennen wir es hier TFNewDF?\n",
    "        \n",
    "\n",
    "        Set_Boundary = setboundary(AllNewDF, StrainWT.genes_df)\n",
    "        \n",
    "        # Defining the model with the two combinations of either\n",
    "        # increasing lower bound (increased forward, decreased reverse reaction)\n",
    "        # decreasing upper bound (decreased forward, increased reverse reaction)\n",
    "        with StrainWT.model as myModel:\n",
    "            # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "\n",
    "#             for Indx in Set_Boundary['lower']:\n",
    "#                 myModel.reactions[Indx].lower_bound = RctNewDF.loc[Indx, 'NewExpr'] * RctNewDF.loc[Indx, 'Expr2Flux']\n",
    "#             # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "#             for Indx in Set_Boundary['upper']:\n",
    "#                 myModel.reactions[Indx].upper_bound = RctNewDF.loc[Indx, 'NewExpr'] * RctNewDF.loc[Indx, 'Expr2Flux']\n",
    "\n",
    "         # \"= ...\" to be replaces by \"= FluxEquation()\"\n",
    "\n",
    "            for Indx in Set_Boundary['lower']:\n",
    "                myModel.reactions[Indx].lower_bound = FluxEquation(Indx, StrainWT.genes_df, AllNewDF)\n",
    "            for Indx in Set_Boundary['upper']:\n",
    "                myModel.reactions[Indx].upper_bound = FluxEquation(Indx, StrainWT.genes_df, AllNewDF)\n",
    "\n",
    "            Fluxes = myModel.optimize()\n",
    "\n",
    "    else:\n",
    "        Fluxes = StrainWT.model.optimize()\n",
    "\n",
    "\n",
    "    return Fluxes.fluxes.values, Fluxes.objective_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FluxValues, FluxObjValue = Help_FluxCalculator ( 'Ecol', wtHost.strain, myHost.strain )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Vergleich von WT und MT aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wtHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mylist = np.array(wtHost.strain.genes_df['Expression'])\n",
    "mylist2 = np.array(myHost.strain.genes_df['Expression'])\n",
    "print('Gene expression differences')\n",
    "np.round(mylist2/mylist,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myl = np.array(wtHost.strain.genes_df['Expr2Flux'])\n",
    "myl2 = np.array(myHost.strain.genes_df['Expr2Flux'])\n",
    "# myHost.strain.genes_df\n",
    "print('Exp2Flux diff')\n",
    "np.round(myl2/myl,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FluxDiff = myHost.strain.genes_df['Fluxes']/wtHost.strain.genes_df['Fluxes']\n",
    "print('Flux reaction differences')\n",
    "np.round(np.nan_to_num(FluxDiff, nan=1, posinf=1, neginf=1),2)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "2001_QuantMiBi1_CobraPy_SimVal.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
