{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup aus MetEngSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "6wxIUGOEReDX"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System ready\n"
     ]
    }
   ],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "from os.path import join\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "from biolabsim import Host, Strain, Ecol\n",
    "from biolabsim import measure_EnzymeLevel1, Help_GenomeGenerator, Strain, make_UpdateExpression, Help_StrainCharacterizer, Help_PromoterStrength, Help_Expr2Flux\n",
    "from Bio.Seq import Seq\n",
    "from pandas.core.frame import DataFrame\n",
    "from copy import deepcopy\n",
    "\n",
    "# import escher\n",
    "# from escher import Builder\n",
    "\n",
    "print('System ready')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Functions defined.\n"
     ]
    }
   ],
   "source": [
    "# define new and changed functions\n",
    "\n",
    "def add_TranscriptionFactor(df:DataFrame, RegID:str, EnzymeTarget):\n",
    "    newdf = df\n",
    "    \n",
    "    # check if TF is already present, add columns to the default TF accordingly, no zero initiation of RegType column\n",
    "    if any('TF_regulated' in s for s in WTdf.columns.values):\n",
    "        TF = {\n",
    "        'RctID': [RegID], \n",
    "        'Expression': [2], # noch zu ändern mit Help_PromoterStrength()?\n",
    "        'Promoter':['GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG'],  # noch zu ändern (echt, random oder Fkt?)\n",
    "        'ORF': ['ATGGAGATGAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "        'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "        'Expr2Flux': [4.5], # noch zu ändern mit Help_Expr2Flux()?\n",
    "        'TF_regulated': [0],\n",
    "        'RegType': 'Hill_Activator'\n",
    "        }\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "        \n",
    "    else:\n",
    "    # add TF row:\n",
    "        Sequence = 'GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG'\n",
    "        TF = {\n",
    "        'RctID': [RegID], \n",
    "        'Expression': Help_PromoterStrength('Ecol', Sequence), # noch zu ändern mit Help_PromoterStrength()?\n",
    "        'Promoter': [Sequence],  # noch zu ändern (echt, random oder Fkt?)\n",
    "        'ORF': ['ATGCCCCCCAAGTAA'], # noch zu ändern (echt, random oder Fkt?)\n",
    "        'Fluxes': [9], # noch zu ändern mit Help_FluxCalculator()?\n",
    "        'Expr2Flux': [4.5] # noch zu ändern mit Help_Expr2Flux()?\n",
    "        }\n",
    "\n",
    "        dfTF = pd.DataFrame(TF)\n",
    "        newdf = newdf.append(dfTF, ignore_index=True)\n",
    "    \n",
    "        # add TF_regulated and RegType column:\n",
    "        zeroarray = np.zeros(len(newdf),dtype=int)\n",
    "        newdf = newdf.assign(TF_regulated=zeroarray) # später zu 0 oder jeweiliger TF-RctID ändern (random oder festgesetzt?)\n",
    "        newdf = newdf.assign(RegType=zeroarray)\n",
    "        \n",
    "        TF_Indx = (newdf['RctID']==RegID) # access TF row\n",
    "        newdf.loc[TF_Indx, 'RegType'] = 'Hill_Activator' # Metabolism Status: TF as activator\n",
    "    \n",
    "    # set Regulation of some Proteins:\n",
    "    for ETarget in EnzymeTarget:\n",
    "    # Future: fct recognizes what substrates are there, then makes TF_act or TF_rep\n",
    "        myIndx = (newdf['RctID']==ETarget)\n",
    "        newdf.loc[myIndx, 'TF_regulated'] = RegID\n",
    "\n",
    "    return newdf\n",
    "\n",
    "\n",
    "def make_DetectRegulatorPromoterMut(StrainWt:Strain, StrainMut:Strain, RctNewDF:DataFrame):\n",
    "#     for obj in (StrainWt, StrainMut):\n",
    "#         assert isinstance(obj, Strain), 'wrong strain type: must be Strain'\n",
    "#     assert isinstance(RctNewDF, DataFrame), 'ReactNewDF of wrong type: input dataframe'\n",
    "    GenesDF = StrainWt.genes_df\n",
    "    # finding regulators in GenesDF\n",
    "    myRegIndx = GenesDF[GenesDF['RegType'] != 0].index.values\n",
    "    # finding index of regulators with changes in the promoter\n",
    "    RegUpdatePromIndx = myRegIndx[RctNewDF.loc[myRegIndx, 'RctFlag'].values]\n",
    "    # finding name of regulators with changes in the promoter\n",
    "    RegUpdatePromName = RctNewDF.loc[RegUpdatePromIndx, 'RctID'].values\n",
    "    # finding enzymes regulated by regulator\n",
    "    EnzUpdateReguIndx = GenesDF['TF_regulated'].isin(RegUpdatePromName)\n",
    "    \n",
    "    return EnzUpdateReguIndx\n",
    "\n",
    "\n",
    "def measure_EnzymeLevel(HostName:str, StrainWT:Strain, StrainMut:Strain):\n",
    "    '''\n",
    "    The function differences of expression levels of enzymes between two strains.\n",
    "    '''\n",
    "    RefGenDF = StrainWT.genes_df\n",
    "    RefGenome = str(StrainWT.genome)\n",
    "    MutGenome = str(StrainMut.genome)\n",
    "    RefModel = StrainWT.model\n",
    "\n",
    "    RctNewDF = Help_StrainCharacterizer(HostName, RefGenDF, RefGenome, MutGenome, RefModel)\n",
    "    \n",
    "    # identifying changes in regulator expression\n",
    "    RegNewAr = make_DetectRegulatorPromoterMut(StrainWT, StrainMut, RctNewDF)\n",
    "\n",
    "    # combining promoter changes enzyme + regulator\n",
    "    AllNewAr = RctNewDF['RctFlag'].values | RegNewAr\n",
    "    AllNewDF = deepcopy(RctNewDF)\n",
    "    AllNewDF['RctFlag'] = AllNewAr  \n",
    "\n",
    "    return AllNewDF\n",
    "\n",
    "\n",
    "def calculate_StandardRegulatedExpression(WTdf):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf['TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "    print (RegTypes)\n",
    "\n",
    "    PreProcFlux = deepcopy(WTdf['Expression'].values)\n",
    "    \n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "        MaxExpr = 2 * WTdf.loc[EnzIndx, 'Expression'].values # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "        \n",
    "        cTF = WTdf.loc[RegIndx, 'Expression'].values # abhängig von geändertem TF-Promotor\n",
    "        K = WTdf.loc[EnzIndx, 'Expression'].values # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95 # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.  \n",
    "#         print ('Eqn: ', RegType, MaxExpr, cTF, K)\n",
    "        \n",
    "        if EquType == 'Hill_Activator':\n",
    "            Flux = MaxExpr / (1+(K/cTF)**n)\n",
    "#             print(Flux)\n",
    "   \n",
    "        elif EquType == 'Hill_Repressor':\n",
    "            Flux = MaxExpr * (1 - 1/ (1+(K/cTF)**n))\n",
    "\n",
    "#         print(EnzIndx)\n",
    "        PreProcFlux[EnzIndx] = Flux\n",
    "\n",
    "    return PreProcFlux\n",
    "\n",
    "    \n",
    "def calculate_mutatedRegulatedExpression(WTdf, AllNewDF):\n",
    "    # finding the different regulator types and their equations which have their promoter changed\n",
    "    RegTypes = [i for i in list(set(WTdf.loc[AllNewDF['RctFlag'], 'TF_regulated'].values))]\n",
    "    # removing the non regulated elements\n",
    "    RegTypes.remove(0)\n",
    "\n",
    "    for RegType in RegTypes:\n",
    "        EnzIndx = WTdf['TF_regulated'] == RegType\n",
    "        RegIndx = WTdf['RctID'] == RegType\n",
    "        EquType = WTdf.loc[WTdf['RctID'] == RegType, 'RegType'].values\n",
    "        MaxExpr = 2 * AllNewDF.loc[EnzIndx, 'RefExpr'].values # bei physiologischem cTF-Wert Flux= mtExpr\n",
    "\n",
    "        cTF = AllNewDF.loc[RegIndx, 'NewExpr'].values # abhängig von geändertem TF-Promotor\n",
    "        K = AllNewDF.loc[EnzIndx, 'RefExpr'].values # physiologischer Wert von c, Expression/Expr2Flux, beeinflusst Aktivität des TF, WT:nicht durch promotor geändert\n",
    "        n = 1.95 # Kim, Harold D.; O'Shea, Erin K. (2008): A quantitative model of transcription factor–activated gene expression. DOI: 10.1038/nsmb.1500.  \n",
    "#         print ('Eqn: ', AllNewDF)\n",
    "        \n",
    "        if EquType == 'Hill_Activator':\n",
    "            corrExpr = MaxExpr / (1+(K/cTF)**n)\n",
    "            \n",
    "        elif EquType == 'Hill_Repressor':\n",
    "            corrExpr = MaxExpr * (1 - 1/ (1+(K/cTF)**n))\n",
    "\n",
    "            \n",
    "        PreProcFlux = AllNewDF['NewExpr'].values\n",
    "        PreProcFlux[EnzIndx] = corrExpr\n",
    "\n",
    "    return PreProcFlux\n",
    "        \n",
    "def setboundary(WTdf, AllNewDF):\n",
    "    '''\n",
    "    Finding and setting the boundaries for cobrapy.\n",
    "    '''\n",
    "\n",
    "    # finding reactions for which the expression has changed\n",
    "    OnlyEnzIndx = WTdf['RegType'] == 0\n",
    "    RctNewDF = AllNewDF.loc[OnlyEnzIndx, ['RctFlag']] == True\n",
    "    RctNew = np.arange((len(RctNewDF)))[np.ravel(RctNewDF.values)]\n",
    "#     OnlyEnzIndx = WTdf[WTdf['RegType'] == 0].index.values\n",
    "#     RctNew = AllNewDF[AllNewDF.loc[OnlyEnzIndx, 'RctFlag'] == True].index.values\n",
    "    \n",
    "    # For reactions with reduced and positive flux: reduce the upper limit,\n",
    "    # For reactions with increased and positive flux: increase the lower limit\n",
    "    # For reactions with negative flux the limits are exchanged.\n",
    "    FluxPos = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux']>0])]\n",
    "    FluxNeg = RctNew[tuple([AllNewDF.loc[RctNew, 'RefFlux']<0])]\n",
    "    # Finding increased and decreased fluxes\n",
    "    FluxInc = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values>1]\n",
    "    FluxDec = RctNew[AllNewDF.loc[RctNew, 'NewFlux'].values / AllNewDF.loc[RctNew, 'RefFlux'].values<1]\n",
    "\n",
    "    # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "    PosIncInd = np.intersect1d(FluxPos,FluxInc)\n",
    "    # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "    PosDecInd = np.intersect1d(FluxPos,FluxDec)\n",
    "    # Comb.3: negative flux with increased expression -> decreasing lower bound\n",
    "    NegIncInd = np.intersect1d(FluxNeg,FluxInc)\n",
    "    # Comb.4: positive flux with increased expression -> increasing lower bound\n",
    "    NegDecInd = np.intersect1d(FluxNeg,FluxDec)\n",
    "\n",
    "    Expr_Change = {'increase': np.hstack([PosIncInd,NegIncInd]),'decrease': np.hstack([PosDecInd,NegDecInd])}\n",
    "    Set_Boundary = {'lower': np.hstack([PosIncInd,NegDecInd]),'upper': np.hstack([PosDecInd,NegIncInd])}\n",
    "    \n",
    "#     Set_Boundary = {'lower': [0,2,3],'upper': []}\n",
    "    return Set_Boundary\n",
    "\n",
    "# Finding reactions to change. Three possibilities exits: 1. the promoter of the enzyme itself has changed, 2. the promoter of a regulator has changed, 3. the enzyme promoter and the regulator promoter have changed\n",
    "# redefining Help_FluxCalculator in simulation/metabolism\n",
    "def Help_FluxCalculator ( HostName:str, Strain:Strain, AllNewDF = None ) :\n",
    "    '''\n",
    "    Calculation of flux values.\n",
    "\n",
    "    This method can work in 2 modes:\n",
    "      [Strain only] : The metabolic model of the WT strain is used for calculation.\n",
    "      [Strain + AllNewDF] : An additional step of resetting boundaries is done on the model\n",
    "        before the fluxes are calculated.\n",
    "\n",
    "    TODO: The \"reset boundary\" step is mutating the `StrainMut.model`, mutation might not be intended.\n",
    "      Because of this, perhaps the other `Strain.model` gets mutated in the process.\n",
    "    '''\n",
    "#     from ..measurement.fluxes import measure_EnzymeLevel1\n",
    "\n",
    "    # adding flux values\n",
    "    # setup of flux boundaries. For the reference boundary changes are set to 'False',\n",
    "    # for mutant strains, ractions with altered promoter sequence will change enzyme levels and boundaries must be changed accordingly, their variable is 'True'\n",
    "\n",
    "    if AllNewDF is not None :\n",
    "        print('resetting boundaries')\n",
    "        # finding reactions for which the expression has changed, and finding the new level\n",
    "        RefFlux = AllNewDF['RefFlux']\n",
    "        NewFlux = AllNewDF['NewFlux']\n",
    "        \n",
    "        Set_Boundary = setboundary(Strain.genes_df, AllNewDF)\n",
    "        # Defining the model with the two combinations of either\n",
    "        # increasing lower bound (increased forward, decreased reverse reaction)\n",
    "        # decreasing upper bound (decreased forward, increased reverse reaction)\n",
    "        with Strain.model as myModel:\n",
    "            # Comb.1: positive flux with increased expression -> increasing lower bound\n",
    "            for Indx in Set_Boundary['lower']:\n",
    "                myModel.reactions[Indx].lower_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "            # Comb.2: positive flux with decreased expression -> decreasing upper bound\n",
    "            for Indx in Set_Boundary['upper']:\n",
    "                myModel.reactions[Indx].upper_bound = AllNewDF.loc[Indx, 'NewFlux']\n",
    "\n",
    "            Fluxes = myModel.optimize()\n",
    "\n",
    "    else:\n",
    "        Fluxes = Strain.model.optimize()\n",
    "\n",
    "\n",
    "    return Fluxes.fluxes.values, Fluxes.objective_value\n",
    "\n",
    "\n",
    "print('Functions defined.')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wild type setup\n",
    "The wild type is generated and transcription factor regulation is added manually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wild type growth rate: 0.87\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TF2', 'TF1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<__array_function__ internals>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>TF_regulated</th>\n",
       "      <th>RegType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>2.579964</td>\n",
       "      <td>GCCCATTGAAGCAGCCGCTCAAGAGGCGTATTTTTGCACG</td>\n",
       "      <td>ATGGACCACCGGTAA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>2.898250</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.435000</td>\n",
       "      <td>GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG</td>\n",
       "      <td>ATGACCCTTCCTTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>4.113402</td>\n",
       "      <td>GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG</td>\n",
       "      <td>ATGTTCTACTATTAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>1.181713</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.799000</td>\n",
       "      <td>GCCCATTGATTGGCGAAGATGAAAAGTATATGGTTGCACG</td>\n",
       "      <td>ATGTGCCTCTACTGA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-20.054476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.996444</td>\n",
       "      <td>GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG</td>\n",
       "      <td>ATGCTTCCCCGGTAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>2.484409</td>\n",
       "      <td>TF2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.332000</td>\n",
       "      <td>GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG</td>\n",
       "      <td>ATGGATCTGATTTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>3.577567</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>1.644000</td>\n",
       "      <td>GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG</td>\n",
       "      <td>ATGCTAGAGTGGTAA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>13.260032</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG</td>\n",
       "      <td>ATGCCGGAACCCTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>9.471972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733000</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG</td>\n",
       "      <td>ATGCCCCCCAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.569859</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TF2</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG</td>\n",
       "      <td>ATGGAGATGAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   RctID  Expression                                  Promoter  \\\n",
       "0    PFK    2.579964  GCCCATTGAAGCAGCCGCTCAAGAGGCGTATTTTTGCACG   \n",
       "1    PFL    1.435000  GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG   \n",
       "2    PGI    4.113402  GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG   \n",
       "3    PGK    0.799000  GCCCATTGATTGGCGAAGATGAAAAGTATATGGTTGCACG   \n",
       "4    PGL    1.996444  GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG   \n",
       "..   ...         ...                                       ...   \n",
       "92  NH4t    1.332000  GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG   \n",
       "93   O2t    1.644000  GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG   \n",
       "94   PDH    0.980000  GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG   \n",
       "95   TF1    5.733000  GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG   \n",
       "96   TF2    2.000000  GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux TF_regulated         RegType  \n",
       "0   ATGGACCACCGGTAA   7.477382   2.898250          TF1               0  \n",
       "1   ATGACCCTTCCTTAA   0.000000   0.000000            0               0  \n",
       "2   ATGTTCTACTATTAA   4.860861   1.181713          TF1               0  \n",
       "3   ATGTGCCTCTACTGA -16.023526 -20.054476            0               0  \n",
       "4   ATGCTTCCCCGGTAA   4.959985   2.484409          TF2               0  \n",
       "..              ...        ...        ...          ...             ...  \n",
       "92  ATGGATCTGATTTAA   4.765319   3.577567            0               0  \n",
       "93  ATGCTAGAGTGGTAA  21.799493  13.260032            0               0  \n",
       "94  ATGCCGGAACCCTAA   9.282533   9.471972            0               0  \n",
       "95  ATGCCCCCCAAGTAA   9.000000   1.569859            0  Hill_Activator  \n",
       "96  ATGGAGATGAAGTAA   9.000000   4.500000            0  Hill_Activator  \n",
       "\n",
       "[97 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wtHost = Ecol()\n",
    "print('Wild type growth rate: {:.2f}'.format(wtHost.strain.objective))\n",
    "\n",
    "# adding regulator TF to dataframe and genome of wt\n",
    "WTdf = wtHost.strain.genes_df\n",
    "WTdf = add_TranscriptionFactor(WTdf, 'TF1', ['PFK','PGI'])\n",
    "WTdf = add_TranscriptionFactor(WTdf, 'TF2', ['PGL'])\n",
    "def Update_DF(wtDF):\n",
    "    # update for regulated expression\n",
    "    corrExpr = calculate_StandardRegulatedExpression(wtDF)\n",
    "    wtDF['Expression'] = corrExpr\n",
    "    # update for Expression to Flux correlation, for all a linear factor\n",
    "    myExpr2Flux = Help_Expr2Flux(wtDF['Fluxes'], corrExpr)\n",
    "    wtDF['Expr2Flux'] = myExpr2Flux\n",
    "    return wtDF\n",
    "# update wildtype dataframe's Expression and Expr2Flux\n",
    "WTdf = Update_DF(WTdf)\n",
    "# putting the genes dataframe back to the wild type\n",
    "wtHost.strain.genes_df = WTdf\n",
    "# Generating new Genome\n",
    "WTGenome_new = Help_GenomeGenerator(WTdf, 500, .6)\n",
    "wtHost.strain.genome = Seq(WTGenome_new)\n",
    "\n",
    "\n",
    "wtHost.strain.genes_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constructing the mutant\n",
    "Mutation targets are manually defined and also the mutations are pre-set. New expression is calculated based on the regulation type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PFK 4\n",
      "PGK 191\n",
      "TF1 5720\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctFlag</th>\n",
       "      <th>RctID</th>\n",
       "      <th>RefExpr</th>\n",
       "      <th>NewExpr</th>\n",
       "      <th>RefFlux</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>NewFlux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>PFK</td>\n",
       "      <td>2.579964</td>\n",
       "      <td>3.211815</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>2.898250</td>\n",
       "      <td>9.308643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>PFL</td>\n",
       "      <td>1.435000</td>\n",
       "      <td>1.435000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>PGI</td>\n",
       "      <td>4.113402</td>\n",
       "      <td>3.282460</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>1.181713</td>\n",
       "      <td>3.878925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>PGK</td>\n",
       "      <td>0.799000</td>\n",
       "      <td>0.782000</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-20.054476</td>\n",
       "      <td>-15.682600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>PGL</td>\n",
       "      <td>1.996444</td>\n",
       "      <td>1.996444</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>2.484409</td>\n",
       "      <td>4.959985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>False</td>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.332000</td>\n",
       "      <td>1.332000</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>3.577567</td>\n",
       "      <td>4.765319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>False</td>\n",
       "      <td>O2t</td>\n",
       "      <td>1.644000</td>\n",
       "      <td>1.644000</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>13.260032</td>\n",
       "      <td>21.799493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>False</td>\n",
       "      <td>PDH</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>9.471972</td>\n",
       "      <td>9.282533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>True</td>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733000</td>\n",
       "      <td>3.334000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.569859</td>\n",
       "      <td>5.233909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>False</td>\n",
       "      <td>TF2</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    RctFlag RctID   RefExpr   NewExpr    RefFlux  Expr2Flux    NewFlux\n",
       "0      True   PFK  2.579964  3.211815   7.477382   2.898250   9.308643\n",
       "1     False   PFL  1.435000  1.435000   0.000000   0.000000   0.000000\n",
       "2      True   PGI  4.113402  3.282460   4.860861   1.181713   3.878925\n",
       "3      True   PGK  0.799000  0.782000 -16.023526 -20.054476 -15.682600\n",
       "4     False   PGL  1.996444  1.996444   4.959985   2.484409   4.959985\n",
       "..      ...   ...       ...       ...        ...        ...        ...\n",
       "92    False  NH4t  1.332000  1.332000   4.765319   3.577567   4.765319\n",
       "93    False   O2t  1.644000  1.644000  21.799493  13.260032  21.799493\n",
       "94    False   PDH  0.980000  0.980000   9.282533   9.471972   9.282533\n",
       "95     True   TF1  5.733000  3.334000   9.000000   1.569859   5.233909\n",
       "96    False   TF2  2.000000  2.000000   9.000000   4.500000   9.000000\n",
       "\n",
       "[97 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generating mutant\n",
    "# We will actually define the mutations by hand and delete the automatically generated genome for now\n",
    "myHost = deepcopy(wtHost)\n",
    "# Mutants to be generated\n",
    "Mutant_Targets = ['PFK','PGK','TF1']\n",
    "\n",
    "for myMutant in Mutant_Targets:\n",
    "#     Mutant_Bool = myHost.strain.genes_df['RctID']==myMutant\n",
    "#     Mutant_Indx = tuple(np.arange(len(Mutant_Bool))[Mutant_Bool])\n",
    "    MTdf = myHost.strain.genes_df.copy()\n",
    "    Mutant_Indx = tuple(MTdf[MTdf['RctID'] == myMutant].index.values)\n",
    "    # Finding the index for later printing\n",
    "    WTIndx = str(wtHost.strain.genome).find(wtHost.strain.genes_df.loc[Mutant_Indx, 'Promoter'])\n",
    "#     # Generating target promoter sequence, ATTGA is always there\n",
    "    PromTar = MTdf.loc[Mutant_Indx, 'Promoter'].replace('ATTGA','CCCCC')\n",
    "    # converting Biopython Seq class to string. This makes string replacements easier\n",
    "    MutGenome = str(myHost.strain.genome)\n",
    "    MutGenome = MutGenome.replace(MTdf.loc[Mutant_Indx, 'Promoter'], PromTar)\n",
    "    myHost.strain.genome = Seq(MutGenome)\n",
    "    MTdf.loc[Mutant_Indx, 'Promoter'] = PromTar\n",
    "    myHost.strain.genes_df = MTdf.copy()\n",
    "#     print('{}, Reference:\\t{},\\nMutated:\\t{}\\nwith index {}'.format(myMutant, wtHost.strain.genome[Mutant_Indx:Mutant_Indx+41], myHost.strain.genome[Mutant_Indx:Mutant_Indx+41], Mutant_Indx))\n",
    "    print(myMutant, WTIndx)\n",
    "\n",
    "# updating expression of regulated enzymes\n",
    "AllNewDF = measure_EnzymeLevel('Ecol', wtHost.strain, myHost.strain)\n",
    "CorrRegExpr = calculate_mutatedRegulatedExpression(WTdf, AllNewDF)\n",
    "AllNewDF['NewExpr'] = CorrRegExpr\n",
    "AllNewDF['NewFlux'] = AllNewDF['NewExpr']*WTdf['Expr2Flux']\n",
    "AllNewDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculation of mutant flux distribution\n",
    "The updated expression activities are converted to reference flux values and according to the reference flux new boundaries are set in the GSMM. The wild type and mutant flux distributions are stored in a dictionary for visualization in Escher."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "resetting boundaries\n"
     ]
    }
   ],
   "source": [
    "solution,_ = Help_FluxCalculator('Ecol', myHost.strain, AllNewDF)\n",
    "RefFluxDict = WTdf.set_index('RctID')['Fluxes'].to_dict()\n",
    "NewFluxDict = {a:b for a,b in zip(WTdf['RctID'].values, solution)}\n",
    "AllFluxDict = [RefFluxDict, NewFluxDict]\n",
    "# {WTdf['RctID'].tolist(): solution}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualization of flux differences\n",
    "Escher is used to visualize the flux differences on the central carbon metabolism map. Open the output-html file `Escher-Metabol-Map.html` by double click in the Jupyter Lab environment and `Trust` the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "-wAeRsXJQI6e"
   },
   "outputs": [],
   "source": [
    "# builder = Builder()\n",
    "# # Load an Escher map\n",
    "# builder.map_name = 'iJO1366.Central metabolism'\n",
    "# builder.model = wtHost.strain.model\n",
    "# builder.reaction_data = AllFluxDict\n",
    "# # Add some data for metabolites\n",
    "# #builder.metabolite_data = solution.shadow_prices\n",
    "# # Simplify the map by hiding some labels\n",
    "# builder.hide_secondary_metabolites = True\n",
    "# #builder.hide_all_labels = True\n",
    "# builder.reaction_scale = [\n",
    "#     { 'type': 'min', 'color': '#000000', 'size': 12 },\n",
    "#     { 'type': 'median', 'color': '#ffffff', 'size': 20 },\n",
    "#     { 'type': 'max', 'color': '#ff0000', 'size': 25 }\n",
    "# ]\n",
    "# builder.reaction_scale_preset = 'GaBuRd'\n",
    "\n",
    "# # Make all the arrows three times as thick\n",
    "# builder.reaction_scale = [\n",
    "#     {k: v * 3 if k == 'size' else v for k, v in x.items()}\n",
    "#     for x in builder.reaction_scale\n",
    "# ]\n",
    "# builder.save_html('Escher-Metabol-Map.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create json file of AllFluxDict to \n",
    "\n",
    "import json\n",
    "\n",
    "# save a single flux vector as JSON\n",
    "flux_dictionary = AllFluxDict\n",
    "filename = 'out'\n",
    "with open(filename + '.json', 'w') as f:\n",
    "    json.dump(flux_dictionary, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>TF_regulated</th>\n",
       "      <th>RegType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>2.579964</td>\n",
       "      <td>GCCCATTGAAGCAGCCGCTCAAGAGGCGTATTTTTGCACG</td>\n",
       "      <td>ATGGACCACCGGTAA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>2.898250</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.435000</td>\n",
       "      <td>GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG</td>\n",
       "      <td>ATGACCCTTCCTTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>4.113402</td>\n",
       "      <td>GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG</td>\n",
       "      <td>ATGTTCTACTATTAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>1.181713</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.799000</td>\n",
       "      <td>GCCCATTGATTGGCGAAGATGAAAAGTATATGGTTGCACG</td>\n",
       "      <td>ATGTGCCTCTACTGA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-20.054476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.996444</td>\n",
       "      <td>GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG</td>\n",
       "      <td>ATGCTTCCCCGGTAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>2.484409</td>\n",
       "      <td>TF2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.332000</td>\n",
       "      <td>GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG</td>\n",
       "      <td>ATGGATCTGATTTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>3.577567</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>1.644000</td>\n",
       "      <td>GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG</td>\n",
       "      <td>ATGCTAGAGTGGTAA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>13.260032</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG</td>\n",
       "      <td>ATGCCGGAACCCTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>9.471972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733000</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG</td>\n",
       "      <td>ATGCCCCCCAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.569859</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TF2</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG</td>\n",
       "      <td>ATGGAGATGAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   RctID  Expression                                  Promoter  \\\n",
       "0    PFK    2.579964  GCCCATTGAAGCAGCCGCTCAAGAGGCGTATTTTTGCACG   \n",
       "1    PFL    1.435000  GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG   \n",
       "2    PGI    4.113402  GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG   \n",
       "3    PGK    0.799000  GCCCATTGATTGGCGAAGATGAAAAGTATATGGTTGCACG   \n",
       "4    PGL    1.996444  GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG   \n",
       "..   ...         ...                                       ...   \n",
       "92  NH4t    1.332000  GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG   \n",
       "93   O2t    1.644000  GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG   \n",
       "94   PDH    0.980000  GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG   \n",
       "95   TF1    5.733000  GCCCATTGACAAGGCTCTCGCGGCCAGGTATAATTGCACG   \n",
       "96   TF2    2.000000  GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux TF_regulated         RegType  \n",
       "0   ATGGACCACCGGTAA   7.477382   2.898250          TF1               0  \n",
       "1   ATGACCCTTCCTTAA   0.000000   0.000000            0               0  \n",
       "2   ATGTTCTACTATTAA   4.860861   1.181713          TF1               0  \n",
       "3   ATGTGCCTCTACTGA -16.023526 -20.054476            0               0  \n",
       "4   ATGCTTCCCCGGTAA   4.959985   2.484409          TF2               0  \n",
       "..              ...        ...        ...          ...             ...  \n",
       "92  ATGGATCTGATTTAA   4.765319   3.577567            0               0  \n",
       "93  ATGCTAGAGTGGTAA  21.799493  13.260032            0               0  \n",
       "94  ATGCCGGAACCCTAA   9.282533   9.471972            0               0  \n",
       "95  ATGCCCCCCAAGTAA   9.000000   1.569859            0  Hill_Activator  \n",
       "96  ATGGAGATGAAGTAA   9.000000   4.500000            0  Hill_Activator  \n",
       "\n",
       "[97 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wtHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RctID</th>\n",
       "      <th>Expression</th>\n",
       "      <th>Promoter</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Fluxes</th>\n",
       "      <th>Expr2Flux</th>\n",
       "      <th>TF_regulated</th>\n",
       "      <th>RegType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>2.579964</td>\n",
       "      <td>GCCCCCCCCAGCAGCCGCTCAAGAGGCGTATTTTTGCACG</td>\n",
       "      <td>ATGGACCACCGGTAA</td>\n",
       "      <td>7.477382</td>\n",
       "      <td>2.898250</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>1.435000</td>\n",
       "      <td>GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG</td>\n",
       "      <td>ATGACCCTTCCTTAA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>4.113402</td>\n",
       "      <td>GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG</td>\n",
       "      <td>ATGTTCTACTATTAA</td>\n",
       "      <td>4.860861</td>\n",
       "      <td>1.181713</td>\n",
       "      <td>TF1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>0.799000</td>\n",
       "      <td>GCCCCCCCCTTGGCGAAGATGAAAAGTATATGGTTGCACG</td>\n",
       "      <td>ATGTGCCTCTACTGA</td>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-20.054476</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>1.996444</td>\n",
       "      <td>GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG</td>\n",
       "      <td>ATGCTTCCCCGGTAA</td>\n",
       "      <td>4.959985</td>\n",
       "      <td>2.484409</td>\n",
       "      <td>TF2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>1.332000</td>\n",
       "      <td>GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG</td>\n",
       "      <td>ATGGATCTGATTTAA</td>\n",
       "      <td>4.765319</td>\n",
       "      <td>3.577567</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>1.644000</td>\n",
       "      <td>GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG</td>\n",
       "      <td>ATGCTAGAGTGGTAA</td>\n",
       "      <td>21.799493</td>\n",
       "      <td>13.260032</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG</td>\n",
       "      <td>ATGCCGGAACCCTAA</td>\n",
       "      <td>9.282533</td>\n",
       "      <td>9.471972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TF1</td>\n",
       "      <td>5.733000</td>\n",
       "      <td>GCCCCCCCCCAAGGCTCTCGCGGCCAGGTATAATTGCACG</td>\n",
       "      <td>ATGCCCCCCAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>1.569859</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TF2</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG</td>\n",
       "      <td>ATGGAGATGAAGTAA</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>4.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>Hill_Activator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   RctID  Expression                                  Promoter  \\\n",
       "0    PFK    2.579964  GCCCCCCCCAGCAGCCGCTCAAGAGGCGTATTTTTGCACG   \n",
       "1    PFL    1.435000  GCCCATTGAGTAGAAATGATACCTTATTTAATTTTGCACG   \n",
       "2    PGI    4.113402  GCCCATTGAGCTACATGTCGGGCAATTGTACAATTGCACG   \n",
       "3    PGK    0.799000  GCCCCCCCCTTGGCGAAGATGAAAAGTATATGGTTGCACG   \n",
       "4    PGL    1.996444  GCCCATTGACGACACATCTGGCCGCTACTACTGTTGCACG   \n",
       "..   ...         ...                                       ...   \n",
       "92  NH4t    1.332000  GCCCATTGATGTACGTGAATAACGAGACTAGACTTGCACG   \n",
       "93   O2t    1.644000  GCCCATTGAATCGCCTTGCTGCAATGCCTAATCTTGCACG   \n",
       "94   PDH    0.980000  GCCCATTGAGTCGTTGATAAGGAATCAATAGCGTTGCACG   \n",
       "95   TF1    5.733000  GCCCCCCCCCAAGGCTCTCGCGGCCAGGTATAATTGCACG   \n",
       "96   TF2    2.000000  GCCCATTGACAAGGCTCTCGCGCGAGTGTATAATTGCACG   \n",
       "\n",
       "                ORF     Fluxes  Expr2Flux TF_regulated         RegType  \n",
       "0   ATGGACCACCGGTAA   7.477382   2.898250          TF1               0  \n",
       "1   ATGACCCTTCCTTAA   0.000000   0.000000            0               0  \n",
       "2   ATGTTCTACTATTAA   4.860861   1.181713          TF1               0  \n",
       "3   ATGTGCCTCTACTGA -16.023526 -20.054476            0               0  \n",
       "4   ATGCTTCCCCGGTAA   4.959985   2.484409          TF2               0  \n",
       "..              ...        ...        ...          ...             ...  \n",
       "92  ATGGATCTGATTTAA   4.765319   3.577567            0               0  \n",
       "93  ATGCTAGAGTGGTAA  21.799493  13.260032            0               0  \n",
       "94  ATGCCGGAACCCTAA   9.282533   9.471972            0               0  \n",
       "95  ATGCCCCCCAAGTAA   9.000000   1.569859            0  Hill_Activator  \n",
       "96  ATGGAGATGAAGTAA   9.000000   4.500000            0  Hill_Activator  \n",
       "\n",
       "[97 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myHost.strain.genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene expression differences\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mylist = np.array(wtHost.strain.genes_df['Expression'])\n",
    "mylist2 = np.array(myHost.strain.genes_df['Expression'])\n",
    "print('Gene expression differences')\n",
    "np.round(mylist2/mylist,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exp2Flux diff\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-11-793ddc7a46d2>:5: RuntimeWarning: invalid value encountered in true_divide\n",
      "  np.round(myl2/myl,2)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 1., nan,  1.,  1.,  1., nan, nan,  1.,  1., nan, nan, nan,  1.,\n",
       "        1.,  1.,  1., nan, nan, nan, nan,  1.,  1., nan,  1.,  1., nan,\n",
       "        1.,  1.,  1.,  1., nan,  1., nan,  1., nan, nan,  1.,  1.,  1.,\n",
       "       nan,  1.,  1.,  1., nan, nan, nan,  1., nan, nan, nan, nan,  1.,\n",
       "       nan, nan,  1.,  1., nan, nan,  1.,  1.,  1., nan, nan,  1., nan,\n",
       "       nan, nan, nan, nan,  1., nan,  1.,  1.,  1.,  1., nan,  1., nan,\n",
       "       nan, nan,  1.,  1.,  1., nan, nan, nan, nan,  1., nan, nan,  1.,\n",
       "       nan,  1.,  1.,  1.,  1.,  1.])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myl = np.array(wtHost.strain.genes_df['Expr2Flux'])\n",
    "myl2 = np.array(myHost.strain.genes_df['Expr2Flux'])\n",
    "# myHost.strain.genes_df\n",
    "print('Exp2Flux diff')\n",
    "np.round(myl2/myl,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Flux reaction differences\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FluxDiff = myHost.strain.genes_df['Fluxes']/wtHost.strain.genes_df['Fluxes']\n",
    "print('Flux reaction differences')\n",
    "np.round(np.nan_to_num(FluxDiff, nan=1, posinf=1, neginf=1),2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.to_csv of 0                          PFK\n",
       "1                          PFL\n",
       "2                          PGI\n",
       "3                          PGK\n",
       "4                          PGL\n",
       "5                        ACALD\n",
       "6                       AKGt2r\n",
       "7                          PGM\n",
       "8                        PIt2r\n",
       "9                       ALCD2x\n",
       "10                      ACALDt\n",
       "11                        ACKr\n",
       "12                         PPC\n",
       "13                      ACONTa\n",
       "14                      ACONTb\n",
       "15                        ATPM\n",
       "16                        PPCK\n",
       "17                       ACt2r\n",
       "18                         PPS\n",
       "19                        ADK1\n",
       "20                       AKGDH\n",
       "21                      ATPS4r\n",
       "22                        PTAr\n",
       "23                         PYK\n",
       "24    BIOMASS_Ecoli_core_w_GAM\n",
       "25                       PYRt2\n",
       "26                        CO2t\n",
       "27                         RPE\n",
       "28                          CS\n",
       "29                         RPI\n",
       "30                    SUCCt2_2\n",
       "31                       CYTBD\n",
       "32                     D_LACt2\n",
       "33                         ENO\n",
       "34                      SUCCt3\n",
       "35                     ETOHt2r\n",
       "36                       SUCDi\n",
       "37                      SUCOAS\n",
       "38                        TALA\n",
       "39                        THD2\n",
       "40                        TKT1\n",
       "41                        TKT2\n",
       "42                         TPI\n",
       "43                     EX_ac_e\n",
       "44                  EX_acald_e\n",
       "45                    EX_akg_e\n",
       "46                    EX_co2_e\n",
       "47                   EX_etoh_e\n",
       "48                    EX_for_e\n",
       "49                    EX_fru_e\n",
       "50                    EX_fum_e\n",
       "51                 EX_glc__D_e\n",
       "52                 EX_gln__L_e\n",
       "53                 EX_glu__L_e\n",
       "54                      EX_h_e\n",
       "55                    EX_h2o_e\n",
       "56                 EX_lac__D_e\n",
       "57                 EX_mal__L_e\n",
       "58                    EX_nh4_e\n",
       "59                     EX_o2_e\n",
       "60                     EX_pi_e\n",
       "61                    EX_pyr_e\n",
       "62                   EX_succ_e\n",
       "63                         FBA\n",
       "64                         FBP\n",
       "65                       FORt2\n",
       "66                        FORt\n",
       "67                        FRD7\n",
       "68                     FRUpts2\n",
       "69                         FUM\n",
       "70                     FUMt2_2\n",
       "71                     G6PDH2r\n",
       "72                        GAPD\n",
       "73                      GLCpts\n",
       "74                        GLNS\n",
       "75                      GLNabc\n",
       "76                       GLUDy\n",
       "77                        GLUN\n",
       "78                       GLUSy\n",
       "79                      GLUt2r\n",
       "80                         GND\n",
       "81                        H2Ot\n",
       "82                      ICDHyr\n",
       "83                         ICL\n",
       "84                       LDH_D\n",
       "85                        MALS\n",
       "86                     MALt2_2\n",
       "87                         MDH\n",
       "88                         ME1\n",
       "89                         ME2\n",
       "90                      NADH16\n",
       "91                     NADTRHD\n",
       "92                        NH4t\n",
       "93                         O2t\n",
       "94                         PDH\n",
       "95                         TF1\n",
       "96                         TF2\n",
       "Name: RctID, dtype: object>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "myHost.strain.genes_df['RctID'].to_csv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "include_colab_link": true,
   "name": "2001_QuantMiBi1_CobraPy_SimVal.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
